// react-daisyui/dist/react-daisyui.modern.js
import { jsxs, jsx, Fragment } from 'react/jsx-runtime';
import React, { forwardRef, cloneElement, useState, useEffect, createRef, useRef, useImperativeHandle, useMemo, useContext } from 'react';

function _extends$1() {
  _extends$1 = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends$1.apply(this, arguments);
}

function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;

  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }

  return target;
}

function toVal(mix) {
	var k, y, str='';

	if (typeof mix === 'string' || typeof mix === 'number') {
		str += mix;
	} else if (typeof mix === 'object') {
		if (Array.isArray(mix)) {
			for (k=0; k < mix.length; k++) {
				if (mix[k]) {
					if (y = toVal(mix[k])) {
						str && (str += ' ');
						str += y;
					}
				}
			}
		} else {
			for (k in mix) {
				if (mix[k]) {
					str && (str += ' ');
					str += k;
				}
			}
		}
	}

	return str;
}

function clsx () {
	var i=0, tmp, x, str='';
	while (i < arguments.length) {
		if (tmp = arguments[i++]) {
			if (x = toVal(tmp)) {
				str && (str += ' ');
				str += x;
			}
		}
	}
	return str;
}

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

var hashlru = function (max) {

  if (!max) throw Error('hashlru must have a max value, of type number, greater than 0')

  var size = 0, cache = Object.create(null), _cache = Object.create(null);

  function update (key, value) {
    cache[key] = value;
    size ++;
    if(size >= max) {
      size = 0;
      _cache = cache;
      cache = Object.create(null);
    }
  }

  return {
    has: function (key) {
      return cache[key] !== undefined || _cache[key] !== undefined
    },
    remove: function (key) {
      if(cache[key] !== undefined)
        cache[key] = undefined;
      if(_cache[key] !== undefined)
        _cache[key] = undefined;
    },
    get: function (key) {
      var v = cache[key];
      if(v !== undefined) return v
      if((v = _cache[key]) !== undefined) {
        update(key, v);
        return v
      }
    },
    set: function (key, value) {
      if(cache[key] !== undefined) cache[key] = value;
      else update(key, value);
    },
    clear: function () {
      cache = Object.create(null);
      _cache = Object.create(null);
    }
  }
};

var HLRU = hashlru;

function getLruCache(cacheSize) {
  if (cacheSize >= 1) {
    return HLRU(cacheSize);
  }

  return {
    get: function get() {
      return undefined;
    },
    set: function set() {}
  };
}

var CLASS_PART_SEPARATOR = '-';
function createClassUtils(config) {
  var classMap = createClassMap(config);

  function getClassGroupId(className) {
    var classParts = className.split(CLASS_PART_SEPARATOR); // Classes like `-inset-1` produce an empty string as first classPart. We assume that classes for negative values are used correctly and remove it from classParts.

    if (classParts[0] === '' && classParts.length !== 1) {
      classParts.shift();
    }

    return getGroupRecursive(classParts, classMap) || getGroupIdForArbitraryProperty(className);
  }

  function getConflictingClassGroupIds(classGroupId) {
    return config.conflictingClassGroups[classGroupId] || [];
  }

  return {
    getClassGroupId: getClassGroupId,
    getConflictingClassGroupIds: getConflictingClassGroupIds
  };
}

function getGroupRecursive(classParts, classPartObject) {
  var _classPartObject$vali;

  if (classParts.length === 0) {
    return classPartObject.classGroupId;
  }

  var currentClassPart = classParts[0];
  var nextClassPartObject = classPartObject.nextPart[currentClassPart];
  var classGroupFromNextClassPart = nextClassPartObject ? getGroupRecursive(classParts.slice(1), nextClassPartObject) : undefined;

  if (classGroupFromNextClassPart) {
    return classGroupFromNextClassPart;
  }

  if (classPartObject.validators.length === 0) {
    return undefined;
  }

  var classRest = classParts.join(CLASS_PART_SEPARATOR);
  return (_classPartObject$vali = classPartObject.validators.find(function (_ref) {
    var validator = _ref.validator;
    return validator(classRest);
  })) == null ? void 0 : _classPartObject$vali.classGroupId;
}

var arbitraryPropertyRegex = /^\[(.+)\]$/;

function getGroupIdForArbitraryProperty(className) {
  if (arbitraryPropertyRegex.test(className)) {
    var arbitraryPropertyClassName = arbitraryPropertyRegex.exec(className)[1];
    var property = arbitraryPropertyClassName == null ? void 0 : arbitraryPropertyClassName.substring(0, arbitraryPropertyClassName.indexOf(':'));

    if (property) {
      // I use two dots here because one dot is used as prefix for class groups in plugins
      return 'arbitrary..' + property;
    }
  }
}
/**
 * Exported for testing only
 */


function createClassMap(config) {
  var theme = config.theme,
      prefix = config.prefix;
  var classMap = {
    nextPart: {},
    validators: []
  };
  var prefixedClassGroupEntries = getPrefixedClassGroupEntries(Object.entries(config.classGroups), prefix);
  prefixedClassGroupEntries.forEach(function (_ref2) {
    var classGroupId = _ref2[0],
        classGroup = _ref2[1];
    processClassesRecursively(classGroup, classMap, classGroupId, theme);
  });
  return classMap;
}

function processClassesRecursively(classGroup, classPartObject, classGroupId, theme) {
  classGroup.forEach(function (classDefinition) {
    if (typeof classDefinition === 'string') {
      var classPartObjectToEdit = classDefinition === '' ? classPartObject : getPart(classPartObject, classDefinition);
      classPartObjectToEdit.classGroupId = classGroupId;
      return;
    }

    if (typeof classDefinition === 'function') {
      if (isThemeGetter(classDefinition)) {
        processClassesRecursively(classDefinition(theme), classPartObject, classGroupId, theme);
        return;
      }

      classPartObject.validators.push({
        validator: classDefinition,
        classGroupId: classGroupId
      });
      return;
    }

    Object.entries(classDefinition).forEach(function (_ref3) {
      var key = _ref3[0],
          classGroup = _ref3[1];
      processClassesRecursively(classGroup, getPart(classPartObject, key), classGroupId, theme);
    });
  });
}

function getPart(classPartObject, path) {
  var currentClassPartObject = classPartObject;
  path.split(CLASS_PART_SEPARATOR).forEach(function (pathPart) {
    if (currentClassPartObject.nextPart[pathPart] === undefined) {
      currentClassPartObject.nextPart[pathPart] = {
        nextPart: {},
        validators: []
      };
    }

    currentClassPartObject = currentClassPartObject.nextPart[pathPart];
  });
  return currentClassPartObject;
}

function isThemeGetter(func) {
  return func.isThemeGetter;
}

function getPrefixedClassGroupEntries(classGroupEntries, prefix) {
  if (!prefix) {
    return classGroupEntries;
  }

  return classGroupEntries.map(function (_ref4) {
    var classGroupId = _ref4[0],
        classGroup = _ref4[1];
    var prefixedClassGroup = classGroup.map(function (classDefinition) {
      if (typeof classDefinition === 'string') {
        return prefix + classDefinition;
      }

      if (typeof classDefinition === 'object') {
        return Object.fromEntries(Object.entries(classDefinition).map(function (_ref5) {
          var key = _ref5[0],
              value = _ref5[1];
          return [prefix + key, value];
        }));
      }

      return classDefinition;
    });
    return [classGroupId, prefixedClassGroup];
  });
}

function createConfigUtils(config) {
  return _extends({
    cache: getLruCache(config.cacheSize)
  }, createClassUtils(config));
}

var SPLIT_CLASSES_REGEX = /\s+/;
var IMPORTANT_MODIFIER = '!'; // Regex is needed so we don't match against colons in labels for arbitrary values like `text-[color:var(--mystery-var)]`
// I'd prefer to use a negative lookbehind for all supported labels, but lookbheinds don't have good browser support yet. More info: https://caniuse.com/js-regexp-lookbehind

var MODIFIER_SEPARATOR_REGEX = /:(?![^[]*\])/;
var MODIFIER_SEPARATOR = ':';
function mergeClassList(classList, configUtils) {
  var getClassGroupId = configUtils.getClassGroupId,
      getConflictingClassGroupIds = configUtils.getConflictingClassGroupIds;
  /**
   * Set of classGroupIds in following format:
   * `{importantModifier}{variantModifiers}{classGroupId}`
   * @example 'float'
   * @example 'hover:focus:bg-color'
   * @example '!md:pr'
   */

  var classGroupsInConflict = new Set();
  return classList.trim().split(SPLIT_CLASSES_REGEX).map(function (originalClassName) {
    var modifiers = originalClassName.split(MODIFIER_SEPARATOR_REGEX);
    var classNameWithImportantModifier = modifiers.pop();
    var hasImportantModifier = classNameWithImportantModifier.startsWith(IMPORTANT_MODIFIER);
    var className = hasImportantModifier ? classNameWithImportantModifier.substring(1) : classNameWithImportantModifier;
    var classGroupId = getClassGroupId(className);

    if (!classGroupId) {
      return {
        isTailwindClass: false,
        originalClassName: originalClassName
      };
    }

    var variantModifier = modifiers.length === 0 ? '' : modifiers.sort().concat('').join(MODIFIER_SEPARATOR);
    var fullModifier = hasImportantModifier ? IMPORTANT_MODIFIER + variantModifier : variantModifier;
    return {
      isTailwindClass: true,
      modifier: fullModifier,
      classGroupId: classGroupId,
      originalClassName: originalClassName
    };
  }).reverse() // Last class in conflict wins, so we need to filter conflicting classes in reverse order.
  .filter(function (parsed) {
    if (!parsed.isTailwindClass) {
      return true;
    }

    var modifier = parsed.modifier,
        classGroupId = parsed.classGroupId;
    var classId = modifier + ":" + classGroupId;

    if (classGroupsInConflict.has(classId)) {
      return false;
    }

    classGroupsInConflict.add(classId);
    getConflictingClassGroupIds(classGroupId).forEach(function (group) {
      return classGroupsInConflict.add(modifier + ":" + group);
    });
    return true;
  }).reverse().map(function (parsed) {
    return parsed.originalClassName;
  }).join(' ');
}

function createTailwindMerge() {
  for (var _len = arguments.length, createConfig = new Array(_len), _key = 0; _key < _len; _key++) {
    createConfig[_key] = arguments[_key];
  }

  var configUtils;
  var cacheGet;
  var cacheSet;
  var functionToCall = initTailwindMerge;

  function initTailwindMerge(classList) {
    var firstCreateConfig = createConfig[0],
        restCreateConfig = createConfig.slice(1);
    var config = restCreateConfig.reduce(function (previousConfig, createConfigCurrent) {
      return createConfigCurrent(previousConfig);
    }, firstCreateConfig());
    configUtils = createConfigUtils(config);
    cacheGet = configUtils.cache.get;
    cacheSet = configUtils.cache.set;
    functionToCall = tailwindMerge;
    return tailwindMerge(classList);
  }

  function tailwindMerge(classList) {
    var cachedResult = cacheGet(classList);

    if (cachedResult) {
      return cachedResult;
    }

    var result = mergeClassList(classList, configUtils);
    cacheSet(classList, result);
    return result;
  }

  return function callTailwindMerge() {
    var classList = '';
    var temp; // Credits â†’ https://github.com/lukeed/clsx/blob/v1.1.1/src/index.js

    for (var index = 0; index < arguments.length; index += 1) {
      if (temp = arguments[index]) {
        classList && (classList += ' ');
        classList += temp;
      }
    }

    return functionToCall(classList);
  };
}

function fromTheme(key) {
  var themeGetter = function themeGetter(theme) {
    return theme[key] || [];
  };

  themeGetter.isThemeGetter = true;
  return themeGetter;
}

var arbitraryValueRegex = /^\[(.+)\]$/;
var fractionRegex = /^\d+\/\d+$/;
var stringLengths = /*#__PURE__*/new Set(['px', 'full', 'screen']);
var tshirtUnitRegex = /^(\d+)?(xs|sm|md|lg|xl)$/;
var lengthUnitRegex = /\d+(%|px|em|rem|vh|vw|pt|pc|in|cm|mm|cap|ch|ex|lh|rlh|vi|vb|vmin|vmax)/; // Shadow always begins with x and y offset separated by underscore

var shadowRegex = /^-?((\d+)?\.?(\d+)[a-z]+|0)_-?((\d+)?\.?(\d+)[a-z]+|0)/;
function isLength(classPart) {
  return !Number.isNaN(Number(classPart)) || stringLengths.has(classPart) || fractionRegex.test(classPart) || isArbitraryLength(classPart);
}
function isArbitraryLength(classPart) {
  var _arbitraryValueRegex$;

  var arbitraryValue = (_arbitraryValueRegex$ = arbitraryValueRegex.exec(classPart)) == null ? void 0 : _arbitraryValueRegex$[1];

  if (arbitraryValue) {
    return arbitraryValue.startsWith('length:') || lengthUnitRegex.test(arbitraryValue);
  }

  return false;
}
function isArbitrarySize(classPart) {
  var _arbitraryValueRegex$2;

  var arbitraryValue = (_arbitraryValueRegex$2 = arbitraryValueRegex.exec(classPart)) == null ? void 0 : _arbitraryValueRegex$2[1];
  return arbitraryValue ? arbitraryValue.startsWith('size:') : false;
}
function isArbitraryPosition(classPart) {
  var _arbitraryValueRegex$3;

  var arbitraryValue = (_arbitraryValueRegex$3 = arbitraryValueRegex.exec(classPart)) == null ? void 0 : _arbitraryValueRegex$3[1];
  return arbitraryValue ? arbitraryValue.startsWith('position:') : false;
}
function isArbitraryUrl(classPart) {
  var _arbitraryValueRegex$4;

  var arbitraryValue = (_arbitraryValueRegex$4 = arbitraryValueRegex.exec(classPart)) == null ? void 0 : _arbitraryValueRegex$4[1];
  return arbitraryValue ? arbitraryValue.startsWith('url(') || arbitraryValue.startsWith('url:') : false;
}
function isArbitraryWeight(classPart) {
  var _arbitraryValueRegex$5;

  var arbitraryValue = (_arbitraryValueRegex$5 = arbitraryValueRegex.exec(classPart)) == null ? void 0 : _arbitraryValueRegex$5[1];
  return arbitraryValue ? !Number.isNaN(Number(arbitraryValue)) || arbitraryValue.startsWith('weight:') : false;
}
function isInteger(classPart) {
  var _arbitraryValueRegex$6;

  var arbitraryValue = (_arbitraryValueRegex$6 = arbitraryValueRegex.exec(classPart)) == null ? void 0 : _arbitraryValueRegex$6[1];

  if (arbitraryValue) {
    return Number.isInteger(Number(arbitraryValue));
  }

  return Number.isInteger(Number(classPart));
}
function isArbitraryValue(classPart) {
  return arbitraryValueRegex.test(classPart);
}
function isAny() {
  return true;
}
function isTshirtSize(classPart) {
  return tshirtUnitRegex.test(classPart);
}
function isArbitraryShadow(classPart) {
  var _arbitraryValueRegex$7;

  var arbitraryValue = (_arbitraryValueRegex$7 = arbitraryValueRegex.exec(classPart)) == null ? void 0 : _arbitraryValueRegex$7[1];

  if (arbitraryValue) {
    return shadowRegex.test(arbitraryValue);
  }

  return false;
}

function getDefaultConfig() {
  var colors = fromTheme('colors');
  var spacing = fromTheme('spacing');
  var blur = fromTheme('blur');
  var brightness = fromTheme('brightness');
  var borderColor = fromTheme('borderColor');
  var borderRadius = fromTheme('borderRadius');
  var borderWidth = fromTheme('borderWidth');
  var contrast = fromTheme('contrast');
  var grayscale = fromTheme('grayscale');
  var hueRotate = fromTheme('hueRotate');
  var invert = fromTheme('invert');
  var gap = fromTheme('gap');
  var gradientColorStops = fromTheme('gradientColorStops');
  var inset = fromTheme('inset');
  var margin = fromTheme('margin');
  var opacity = fromTheme('opacity');
  var padding = fromTheme('padding');
  var saturate = fromTheme('saturate');
  var scale = fromTheme('scale');
  var sepia = fromTheme('sepia');
  var skew = fromTheme('skew');
  var space = fromTheme('space');
  var translate = fromTheme('translate');

  var getOverscroll = function getOverscroll() {
    return ['auto', 'contain', 'none'];
  };

  var getOverflow = function getOverflow() {
    return ['auto', 'hidden', 'clip', 'visible', 'scroll'];
  };

  var getSpacingWithAuto = function getSpacingWithAuto() {
    return ['auto', spacing];
  };

  var getLengthWithEmpty = function getLengthWithEmpty() {
    return ['', isLength];
  };

  var getIntegerWithAuto = function getIntegerWithAuto() {
    return ['auto', isInteger];
  };

  var getPositions = function getPositions() {
    return ['bottom', 'center', 'left', 'left-bottom', 'left-top', 'right', 'right-bottom', 'right-top', 'top'];
  };

  var getLineStyles = function getLineStyles() {
    return ['solid', 'dashed', 'dotted', 'double', 'none'];
  };

  var getBlendModes = function getBlendModes() {
    return ['normal', 'multiply', 'screen', 'overlay', 'darken', 'lighten', 'color-dodge', 'color-burn', 'hard-light', 'soft-light', 'difference', 'exclusion', 'hue', 'saturation', 'color', 'luminosity'];
  };

  var getAlign = function getAlign() {
    return ['start', 'end', 'center', 'between', 'around', 'evenly'];
  };

  var getZeroAndEmpty = function getZeroAndEmpty() {
    return ['', '0', isArbitraryValue];
  };

  var getBreaks = function getBreaks() {
    return ['auto', 'avoid', 'all', 'avoid-page', 'page', 'left', 'right', 'column'];
  };

  return {
    cacheSize: 500,
    theme: {
      colors: [isAny],
      spacing: [isLength],
      blur: ['none', '', isTshirtSize, isArbitraryLength],
      brightness: [isInteger],
      borderColor: [colors],
      borderRadius: ['none', '', 'full', isTshirtSize, isArbitraryLength],
      borderWidth: getLengthWithEmpty(),
      contrast: [isInteger],
      grayscale: getZeroAndEmpty(),
      hueRotate: [isInteger],
      invert: getZeroAndEmpty(),
      gap: [spacing],
      gradientColorStops: [colors],
      inset: getSpacingWithAuto(),
      margin: getSpacingWithAuto(),
      opacity: [isInteger],
      padding: [spacing],
      saturate: [isInteger],
      scale: [isInteger],
      sepia: getZeroAndEmpty(),
      skew: [isInteger, isArbitraryValue],
      space: [spacing],
      translate: [spacing]
    },
    classGroups: {
      // Layout

      /**
       * Aspect Ratio
       * @see https://tailwindcss.com/docs/aspect-ratio
       */
      aspect: [{
        aspect: ['auto', 'square', 'video', isArbitraryValue]
      }],

      /**
       * Container
       * @see https://tailwindcss.com/docs/container
       */
      container: ['container'],

      /**
       * Columns
       * @see https://tailwindcss.com/docs/columns
       */
      columns: [{
        columns: [isTshirtSize]
      }],

      /**
       * Break After
       * @see https://tailwindcss.com/docs/break-after
       */
      'break-after': [{
        'break-after': getBreaks()
      }],

      /**
       * Break Before
       * @see https://tailwindcss.com/docs/break-before
       */
      'break-before': [{
        'break-before': getBreaks()
      }],

      /**
       * Break Inside
       * @see https://tailwindcss.com/docs/break-inside
       */
      'break-inside': [{
        'break-inside': ['auto', 'avoid', 'avoid-page', 'avoid-column']
      }],

      /**
       * Box Decoration Break
       * @see https://tailwindcss.com/docs/box-decoration-break
       */
      'box-decoration': [{
        'box-decoration': ['slice', 'clone']
      }],

      /**
       * Box Sizing
       * @see https://tailwindcss.com/docs/box-sizing
       */
      box: [{
        box: ['border', 'content']
      }],

      /**
       * Display
       * @see https://tailwindcss.com/docs/display
       */
      display: ['block', 'inline-block', 'inline', 'flex', 'inline-flex', 'table', 'inline-table', 'table-caption', 'table-cell', 'table-column', 'table-column-group', 'table-footer-group', 'table-header-group', 'table-row-group', 'table-row', 'flow-root', 'grid', 'inline-grid', 'contents', 'list-item', 'hidden'],

      /**
       * Floats
       * @see https://tailwindcss.com/docs/float
       */
      "float": [{
        "float": ['right', 'left', 'none']
      }],

      /**
       * Clear
       * @see https://tailwindcss.com/docs/clear
       */
      clear: [{
        clear: ['left', 'right', 'both', 'none']
      }],

      /**
       * Isolation
       * @see https://tailwindcss.com/docs/isolation
       */
      isolation: ['isolate', 'isolation-auto'],

      /**
       * Object Fit
       * @see https://tailwindcss.com/docs/object-fit
       */
      'object-fit': [{
        object: ['contain', 'cover', 'fill', 'none', 'scale-down']
      }],

      /**
       * Object Position
       * @see https://tailwindcss.com/docs/object-position
       */
      'object-position': [{
        object: [].concat(getPositions(), [isArbitraryValue])
      }],

      /**
       * Overflow
       * @see https://tailwindcss.com/docs/overflow
       */
      overflow: [{
        overflow: getOverflow()
      }],

      /**
       * Overflow X
       * @see https://tailwindcss.com/docs/overflow
       */
      'overflow-x': [{
        'overflow-x': getOverflow()
      }],

      /**
       * Overflow Y
       * @see https://tailwindcss.com/docs/overflow
       */
      'overflow-y': [{
        'overflow-y': getOverflow()
      }],

      /**
       * Overscroll Behavior
       * @see https://tailwindcss.com/docs/overscroll-behavior
       */
      overscroll: [{
        overscroll: getOverscroll()
      }],

      /**
       * Overscroll Behavior X
       * @see https://tailwindcss.com/docs/overscroll-behavior
       */
      'overscroll-x': [{
        'overscroll-x': getOverscroll()
      }],

      /**
       * Overscroll Behavior Y
       * @see https://tailwindcss.com/docs/overscroll-behavior
       */
      'overscroll-y': [{
        'overscroll-y': getOverscroll()
      }],

      /**
       * Position
       * @see https://tailwindcss.com/docs/position
       */
      position: ['static', 'fixed', 'absolute', 'relative', 'sticky'],

      /**
       * Top / Right / Bottom / Left
       * @see https://tailwindcss.com/docs/top-right-bottom-left
       */
      inset: [{
        inset: [inset]
      }],

      /**
       * Right / Left
       * @see https://tailwindcss.com/docs/top-right-bottom-left
       */
      'inset-x': [{
        'inset-x': [inset]
      }],

      /**
       * Top / Bottom
       * @see https://tailwindcss.com/docs/top-right-bottom-left
       */
      'inset-y': [{
        'inset-y': [inset]
      }],

      /**
       * Top
       * @see https://tailwindcss.com/docs/top-right-bottom-left
       */
      top: [{
        top: [inset]
      }],

      /**
       * Right
       * @see https://tailwindcss.com/docs/top-right-bottom-left
       */
      right: [{
        right: [inset]
      }],

      /**
       * Bottom
       * @see https://tailwindcss.com/docs/top-right-bottom-left
       */
      bottom: [{
        bottom: [inset]
      }],

      /**
       * Left
       * @see https://tailwindcss.com/docs/top-right-bottom-left
       */
      left: [{
        left: [inset]
      }],

      /**
       * Visibility
       * @see https://tailwindcss.com/docs/visibility
       */
      visibility: ['visible', 'invisible'],

      /**
       * Z-Index
       * @see https://tailwindcss.com/docs/z-index
       */
      z: [{
        z: [isLength]
      }],
      // Flexbox and Grid

      /**
       * Flex Basis
       * @see https://tailwindcss.com/docs/flex-basis
       */
      basis: [{
        basis: [spacing]
      }],

      /**
       * Flex Direction
       * @see https://tailwindcss.com/docs/flex-direction
       */
      'flex-direction': [{
        flex: ['row', 'row-reverse', 'col', 'col-reverse']
      }],

      /**
       * Flex Wrap
       * @see https://tailwindcss.com/docs/flex-wrap
       */
      'flex-wrap': [{
        flex: ['wrap', 'wrap-reverse', 'nowrap']
      }],

      /**
       * Flex
       * @see https://tailwindcss.com/docs/flex
       */
      flex: [{
        flex: ['1', 'auto', 'initial', 'none', isArbitraryValue]
      }],

      /**
       * Flex Grow
       * @see https://tailwindcss.com/docs/flex-grow
       */
      grow: [{
        grow: getZeroAndEmpty()
      }],

      /**
       * Flex Shrink
       * @see https://tailwindcss.com/docs/flex-shrink
       */
      shrink: [{
        shrink: getZeroAndEmpty()
      }],

      /**
       * Order
       * @see https://tailwindcss.com/docs/order
       */
      order: [{
        order: ['first', 'last', 'none', isInteger]
      }],

      /**
       * Grid Template Columns
       * @see https://tailwindcss.com/docs/grid-template-columns
       */
      'grid-cols': [{
        'grid-cols': [isAny]
      }],

      /**
       * Grid Column Start / End
       * @see https://tailwindcss.com/docs/grid-column
       */
      'col-start-end': [{
        col: ['auto', {
          span: [isInteger]
        }]
      }],

      /**
       * Grid Column Start
       * @see https://tailwindcss.com/docs/grid-column
       */
      'col-start': [{
        'col-start': getIntegerWithAuto()
      }],

      /**
       * Grid Column End
       * @see https://tailwindcss.com/docs/grid-column
       */
      'col-end': [{
        'col-end': getIntegerWithAuto()
      }],

      /**
       * Grid Template Rows
       * @see https://tailwindcss.com/docs/grid-template-rows
       */
      'grid-rows': [{
        'grid-rows': [isAny]
      }],

      /**
       * Grid Row Start / End
       * @see https://tailwindcss.com/docs/grid-row
       */
      'row-start-end': [{
        row: ['auto', {
          span: [isInteger]
        }]
      }],

      /**
       * Grid Row Start
       * @see https://tailwindcss.com/docs/grid-row
       */
      'row-start': [{
        'row-start': getIntegerWithAuto()
      }],

      /**
       * Grid Row End
       * @see https://tailwindcss.com/docs/grid-row
       */
      'row-end': [{
        'row-end': getIntegerWithAuto()
      }],

      /**
       * Grid Auto Flow
       * @see https://tailwindcss.com/docs/grid-auto-flow
       */
      'grid-flow': [{
        'grid-flow': ['row', 'col', 'row-dense', 'col-dense']
      }],

      /**
       * Grid Auto Columns
       * @see https://tailwindcss.com/docs/grid-auto-columns
       */
      'auto-cols': [{
        'auto-cols': ['auto', 'min', 'max', 'fr', isArbitraryValue]
      }],

      /**
       * Grid Auto Rows
       * @see https://tailwindcss.com/docs/grid-auto-rows
       */
      'auto-rows': [{
        'auto-rows': ['auto', 'min', 'max', 'fr', isArbitraryValue]
      }],

      /**
       * Gap
       * @see https://tailwindcss.com/docs/gap
       */
      gap: [{
        gap: [gap]
      }],

      /**
       * Gap X
       * @see https://tailwindcss.com/docs/gap
       */
      'gap-x': [{
        'gap-x': [gap]
      }],

      /**
       * Gap Y
       * @see https://tailwindcss.com/docs/gap
       */
      'gap-y': [{
        'gap-y': [gap]
      }],

      /**
       * Justify Content
       * @see https://tailwindcss.com/docs/justify-content
       */
      'justify-content': [{
        justify: getAlign()
      }],

      /**
       * Justify Items
       * @see https://tailwindcss.com/docs/justify-items
       */
      'justify-items': [{
        'justify-items': ['start', 'end', 'center', 'stretch']
      }],

      /**
       * Justify Self
       * @see https://tailwindcss.com/docs/justify-self
       */
      'justify-self': [{
        'justify-self': ['auto', 'start', 'end', 'center', 'stretch']
      }],

      /**
       * Align Content
       * @see https://tailwindcss.com/docs/align-content
       */
      'align-content': [{
        content: getAlign()
      }],

      /**
       * Align Items
       * @see https://tailwindcss.com/docs/align-items
       */
      'align-items': [{
        items: ['start', 'end', 'center', 'baseline', 'stretch']
      }],

      /**
       * Align Self
       * @see https://tailwindcss.com/docs/align-self
       */
      'align-self': [{
        self: ['auto', 'start', 'end', 'center', 'stretch', 'baseline']
      }],

      /**
       * Place Content
       * @see https://tailwindcss.com/docs/place-content
       */
      'place-content': [{
        'place-content': [].concat(getAlign(), ['stretch'])
      }],

      /**
       * Place Items
       * @see https://tailwindcss.com/docs/place-items
       */
      'place-items': [{
        'place-items': ['start', 'end', 'center', 'stretch']
      }],

      /**
       * Place Self
       * @see https://tailwindcss.com/docs/place-self
       */
      'place-self': [{
        'place-self': ['auto', 'start', 'end', 'center', 'stretch']
      }],
      // Spacing

      /**
       * Padding
       * @see https://tailwindcss.com/docs/padding
       */
      p: [{
        p: [padding]
      }],

      /**
       * Padding X
       * @see https://tailwindcss.com/docs/padding
       */
      px: [{
        px: [padding]
      }],

      /**
       * Padding Y
       * @see https://tailwindcss.com/docs/padding
       */
      py: [{
        py: [padding]
      }],

      /**
       * Padding Top
       * @see https://tailwindcss.com/docs/padding
       */
      pt: [{
        pt: [padding]
      }],

      /**
       * Padding Right
       * @see https://tailwindcss.com/docs/padding
       */
      pr: [{
        pr: [padding]
      }],

      /**
       * Padding Bottom
       * @see https://tailwindcss.com/docs/padding
       */
      pb: [{
        pb: [padding]
      }],

      /**
       * Padding Left
       * @see https://tailwindcss.com/docs/padding
       */
      pl: [{
        pl: [padding]
      }],

      /**
       * Margin
       * @see https://tailwindcss.com/docs/margin
       */
      m: [{
        m: [margin]
      }],

      /**
       * Margin X
       * @see https://tailwindcss.com/docs/margin
       */
      mx: [{
        mx: [margin]
      }],

      /**
       * Margin Y
       * @see https://tailwindcss.com/docs/margin
       */
      my: [{
        my: [margin]
      }],

      /**
       * Margin Top
       * @see https://tailwindcss.com/docs/margin
       */
      mt: [{
        mt: [margin]
      }],

      /**
       * Margin Right
       * @see https://tailwindcss.com/docs/margin
       */
      mr: [{
        mr: [margin]
      }],

      /**
       * Margin Bottom
       * @see https://tailwindcss.com/docs/margin
       */
      mb: [{
        mb: [margin]
      }],

      /**
       * Margin Left
       * @see https://tailwindcss.com/docs/margin
       */
      ml: [{
        ml: [margin]
      }],

      /**
       * Space Between X
       * @see https://tailwindcss.com/docs/space
       */
      'space-x': [{
        'space-x': [space]
      }],

      /**
       * Space Between X Reverse
       * @see https://tailwindcss.com/docs/space
       */
      'space-x-reverse': ['space-x-reverse'],

      /**
       * Space Between Y
       * @see https://tailwindcss.com/docs/space
       */
      'space-y': [{
        'space-y': [space]
      }],

      /**
       * Space Between Y Reverse
       * @see https://tailwindcss.com/docs/space
       */
      'space-y-reverse': ['space-y-reverse'],
      // Sizing

      /**
       * Width
       * @see https://tailwindcss.com/docs/width
       */
      w: [{
        w: ['auto', 'min', 'max', spacing]
      }],

      /**
       * Min-Width
       * @see https://tailwindcss.com/docs/min-width
       */
      'min-w': [{
        'min-w': ['min', 'max', 'fit', isLength]
      }],

      /**
       * Max-Width
       * @see https://tailwindcss.com/docs/max-width
       */
      'max-w': [{
        'max-w': ['0', 'none', 'full', 'min', 'max', 'fit', 'prose', {
          screen: [isTshirtSize]
        }, isTshirtSize, isArbitraryLength]
      }],

      /**
       * Height
       * @see https://tailwindcss.com/docs/height
       */
      h: [{
        h: getSpacingWithAuto()
      }],

      /**
       * Min-Height
       * @see https://tailwindcss.com/docs/min-height
       */
      'min-h': [{
        'min-h': ['min', 'max', 'fit', isLength]
      }],

      /**
       * Max-Height
       * @see https://tailwindcss.com/docs/max-height
       */
      'max-h': [{
        'max-h': [spacing, 'min', 'max', 'fit']
      }],
      // Typography

      /**
       * Font Size
       * @see https://tailwindcss.com/docs/font-size
       */
      'font-size': [{
        text: ['base', isTshirtSize, isArbitraryLength]
      }],

      /**
       * Font Smoothing
       * @see https://tailwindcss.com/docs/font-smoothing
       */
      'font-smoothing': ['antialiased', 'subpixel-antialiased'],

      /**
       * Font Style
       * @see https://tailwindcss.com/docs/font-style
       */
      'font-style': ['italic', 'not-italic'],

      /**
       * Font Weight
       * @see https://tailwindcss.com/docs/font-weight
       */
      'font-weight': [{
        font: ['thin', 'extralight', 'light', 'normal', 'medium', 'semibold', 'bold', 'extrabold', 'black', isArbitraryWeight]
      }],

      /**
       * Font Family
       * @see https://tailwindcss.com/docs/font-family
       */
      'font-family': [{
        font: [isAny]
      }],

      /**
       * Font Variant Numeric
       * @see https://tailwindcss.com/docs/font-variant-numeric
       */
      'fvn-normal': ['normal-nums'],

      /**
       * Font Variant Numeric
       * @see https://tailwindcss.com/docs/font-variant-numeric
       */
      'fvn-ordinal': ['ordinal'],

      /**
       * Font Variant Numeric
       * @see https://tailwindcss.com/docs/font-variant-numeric
       */
      'fvn-slashed-zero': ['slashed-zero'],

      /**
       * Font Variant Numeric
       * @see https://tailwindcss.com/docs/font-variant-numeric
       */
      'fvn-figure': ['lining-nums', 'oldstyle-nums'],

      /**
       * Font Variant Numeric
       * @see https://tailwindcss.com/docs/font-variant-numeric
       */
      'fvn-spacing': ['proportional-nums', 'tabular-nums'],

      /**
       * Font Variant Numeric
       * @see https://tailwindcss.com/docs/font-variant-numeric
       */
      'fvn-fraction': ['diagonal-fractions', 'stacked-fractons'],

      /**
       * Letter Spacing
       * @see https://tailwindcss.com/docs/letter-spacing
       */
      tracking: [{
        tracking: ['tighter', 'tight', 'normal', 'wide', 'wider', 'widest', isArbitraryLength]
      }],

      /**
       * Line Height
       * @see https://tailwindcss.com/docs/line-height
       */
      leading: [{
        leading: ['none', 'tight', 'snug', 'normal', 'relaxed', 'loose', isLength]
      }],

      /**
       * List Style Type
       * @see https://tailwindcss.com/docs/list-style-type
       */
      'list-style-type': [{
        list: ['none', 'disc', 'decimal', isArbitraryValue]
      }],

      /**
       * List Style Position
       * @see https://tailwindcss.com/docs/list-style-position
       */
      'list-style-position': [{
        list: ['inside', 'outside']
      }],

      /**
       * Placeholder Color
       * @deprecated since Tailwind CSS v3.0.0
       * @see https://tailwindcss.com/docs/placeholder-color
       */
      'placeholder-color': [{
        placeholder: [colors]
      }],

      /**
       * Placeholder Opacity
       * @see https://tailwindcss.com/docs/placeholder-opacity
       */
      'placeholder-opacity': [{
        'placeholder-opacity': [opacity]
      }],

      /**
       * Text Alignment
       * @see https://tailwindcss.com/docs/text-align
       */
      'text-alignment': [{
        text: ['left', 'center', 'right', 'justify']
      }],

      /**
       * Text Color
       * @see https://tailwindcss.com/docs/text-color
       */
      'text-color': [{
        text: [colors]
      }],

      /**
       * Text Opacity
       * @see https://tailwindcss.com/docs/text-opacity
       */
      'text-opacity': [{
        'text-opacity': [opacity]
      }],

      /**
       * Text Decoration
       * @see https://tailwindcss.com/docs/text-decoration
       */
      'text-decoration': ['underline', 'overline', 'line-through', 'no-underline'],

      /**
       * Text Decoration Style
       * @see https://tailwindcss.com/docs/text-decoration-style
       */
      'text-decoration-style': [{
        decoration: [].concat(getLineStyles(), ['wavy'])
      }],

      /**
       * Text Decoration Thickness
       * @see https://tailwindcss.com/docs/text-decoration-thickness
       */
      'text-decoration-thickness': [{
        decoration: ['auto', 'from-font', isLength]
      }],

      /**
       * Text Underline Offset
       * @see https://tailwindcss.com/docs/text-underline-offset
       */
      'underline-offset': [{
        'underline-offset': ['auto', isLength]
      }],

      /**
       * Text Decoration Color
       * @see https://tailwindcss.com/docs/text-decoration-color
       */
      'text-decoration-color': [{
        decoration: [colors]
      }],

      /**
       * Text Transform
       * @see https://tailwindcss.com/docs/text-transform
       */
      'text-transform': ['uppercase', 'lowercase', 'capitalize', 'normal-case'],

      /**
       * Text Overflow
       * @see https://tailwindcss.com/docs/text-overflow
       */
      'text-overflow': ['truncate', 'text-ellipsis', 'text-clip'],

      /**
       * Text Indent
       * @see https://tailwindcss.com/docs/text-indent
       */
      indent: [{
        indent: [spacing]
      }],

      /**
       * Vertical Alignment
       * @see https://tailwindcss.com/docs/vertical-align
       */
      'vertical-align': [{
        align: ['baseline', 'top', 'middle', 'bottom', 'text-top', 'text-bottom', 'sub', 'super', isArbitraryLength]
      }],

      /**
       * Whitespace
       * @see https://tailwindcss.com/docs/whitespace
       */
      whitespace: [{
        whitespace: ['normal', 'nowrap', 'pre', 'pre-line', 'pre-wrap']
      }],

      /**
       * Word Break
       * @see https://tailwindcss.com/docs/word-break
       */
      "break": [{
        "break": ['normal', 'words', 'all']
      }],

      /**
       * Content
       * @see https://tailwindcss.com/docs/content
       */
      content: [{
        content: ['none', isArbitraryValue]
      }],
      // Backgrounds

      /**
       * Background Attachment
       * @see https://tailwindcss.com/docs/background-attachment
       */
      'bg-attachment': [{
        bg: ['fixed', 'local', 'scroll']
      }],

      /**
       * Background Clip
       * @see https://tailwindcss.com/docs/background-clip
       */
      'bg-clip': [{
        'bg-clip': ['border', 'padding', 'content', 'text']
      }],

      /**
       * Background Opacity
       * @deprecated since Tailwind CSS v3.0.0
       * @see https://tailwindcss.com/docs/background-opacity
       */
      'bg-opacity': [{
        'bg-opacity': [opacity]
      }],

      /**
       * Background Origin
       * @see https://tailwindcss.com/docs/background-origin
       */
      'bg-origin': [{
        'bg-origin': ['border', 'padding', 'content']
      }],

      /**
       * Background Position
       * @see https://tailwindcss.com/docs/background-position
       */
      'bg-position': [{
        bg: [].concat(getPositions(), [isArbitraryPosition])
      }],

      /**
       * Background Repeat
       * @see https://tailwindcss.com/docs/background-repeat
       */
      'bg-repeat': [{
        bg: ['no-repeat', {
          repeat: ['', 'x', 'y', 'round', 'space']
        }]
      }],

      /**
       * Background Size
       * @see https://tailwindcss.com/docs/background-size
       */
      'bg-size': [{
        bg: ['auto', 'cover', 'contain', isArbitrarySize]
      }],

      /**
       * Background Image
       * @see https://tailwindcss.com/docs/background-image
       */
      'bg-image': [{
        bg: ['none', {
          'gradient-to': ['t', 'tr', 'r', 'br', 'b', 'bl', 'l', 'tl']
        }, isArbitraryUrl]
      }],

      /**
       * Background Color
       * @see https://tailwindcss.com/docs/background-color
       */
      'bg-color': [{
        bg: [colors]
      }],

      /**
       * Gradient Color Stops From
       * @see https://tailwindcss.com/docs/gradient-color-stops
       */
      'gradient-from': [{
        from: [gradientColorStops]
      }],

      /**
       * Gradient Color Stops Via
       * @see https://tailwindcss.com/docs/gradient-color-stops
       */
      'gradient-via': [{
        via: [gradientColorStops]
      }],

      /**
       * Gradient Color Stops To
       * @see https://tailwindcss.com/docs/gradient-color-stops
       */
      'gradient-to': [{
        to: [gradientColorStops]
      }],
      // Borders

      /**
       * Border Radius
       * @see https://tailwindcss.com/docs/border-radius
       */
      rounded: [{
        rounded: [borderRadius]
      }],

      /**
       * Border Radius Top
       * @see https://tailwindcss.com/docs/border-radius
       */
      'rounded-t': [{
        'rounded-t': [borderRadius]
      }],

      /**
       * Border Radius Right
       * @see https://tailwindcss.com/docs/border-radius
       */
      'rounded-r': [{
        'rounded-r': [borderRadius]
      }],

      /**
       * Border Radius Bottom
       * @see https://tailwindcss.com/docs/border-radius
       */
      'rounded-b': [{
        'rounded-b': [borderRadius]
      }],

      /**
       * Border Radius Left
       * @see https://tailwindcss.com/docs/border-radius
       */
      'rounded-l': [{
        'rounded-l': [borderRadius]
      }],

      /**
       * Border Radius Top Left
       * @see https://tailwindcss.com/docs/border-radius
       */
      'rounded-tl': [{
        'rounded-tl': [borderRadius]
      }],

      /**
       * Border Radius Top Right
       * @see https://tailwindcss.com/docs/border-radius
       */
      'rounded-tr': [{
        'rounded-tr': [borderRadius]
      }],

      /**
       * Border Radius Bottom Right
       * @see https://tailwindcss.com/docs/border-radius
       */
      'rounded-br': [{
        'rounded-br': [borderRadius]
      }],

      /**
       * Border Radius Bottom Left
       * @see https://tailwindcss.com/docs/border-radius
       */
      'rounded-bl': [{
        'rounded-bl': [borderRadius]
      }],

      /**
       * Border Width
       * @see https://tailwindcss.com/docs/border-width
       */
      'border-w': [{
        border: [borderWidth]
      }],

      /**
       * Border Width X
       * @see https://tailwindcss.com/docs/border-width
       */
      'border-w-x': [{
        'border-x': [borderWidth]
      }],

      /**
       * Border Width Y
       * @see https://tailwindcss.com/docs/border-width
       */
      'border-w-y': [{
        'border-y': [borderWidth]
      }],

      /**
       * Border Width Top
       * @see https://tailwindcss.com/docs/border-width
       */
      'border-w-t': [{
        'border-t': [borderWidth]
      }],

      /**
       * Border Width Right
       * @see https://tailwindcss.com/docs/border-width
       */
      'border-w-r': [{
        'border-r': [borderWidth]
      }],

      /**
       * Border Width Bottom
       * @see https://tailwindcss.com/docs/border-width
       */
      'border-w-b': [{
        'border-b': [borderWidth]
      }],

      /**
       * Border Width Left
       * @see https://tailwindcss.com/docs/border-width
       */
      'border-w-l': [{
        'border-l': [borderWidth]
      }],

      /**
       * Border Opacity
       * @see https://tailwindcss.com/docs/border-opacity
       */
      'border-opacity': [{
        'border-opacity': [opacity]
      }],

      /**
       * Border Style
       * @see https://tailwindcss.com/docs/border-style
       */
      'border-style': [{
        border: [].concat(getLineStyles(), ['hidden'])
      }],

      /**
       * Divide Width X
       * @see https://tailwindcss.com/docs/divide-width
       */
      'divide-x': [{
        'divide-x': [borderWidth]
      }],

      /**
       * Divide Width X Reverse
       * @see https://tailwindcss.com/docs/divide-width
       */
      'divide-x-reverse': ['divide-x-reverse'],

      /**
       * Divide Width Y
       * @see https://tailwindcss.com/docs/divide-width
       */
      'divide-y': [{
        'divide-y': [borderWidth]
      }],

      /**
       * Divide Width Y Reverse
       * @see https://tailwindcss.com/docs/divide-width
       */
      'divide-y-reverse': ['divide-y-reverse'],

      /**
       * Divide Opacity
       * @see https://tailwindcss.com/docs/divide-opacity
       */
      'divide-opacity': [{
        'divide-opacity': [opacity]
      }],

      /**
       * Divide Style
       * @see https://tailwindcss.com/docs/divide-style
       */
      'divide-style': [{
        divide: getLineStyles()
      }],

      /**
       * Border Color
       * @see https://tailwindcss.com/docs/border-color
       */
      'border-color': [{
        border: [borderColor]
      }],

      /**
       * Border Color X
       * @see https://tailwindcss.com/docs/border-color
       */
      'border-color-x': [{
        'border-x': [borderColor]
      }],

      /**
       * Border Color Y
       * @see https://tailwindcss.com/docs/border-color
       */
      'border-color-y': [{
        'border-y': [borderColor]
      }],

      /**
       * Border Color Top
       * @see https://tailwindcss.com/docs/border-color
       */
      'border-color-t': [{
        'border-t': [borderColor]
      }],

      /**
       * Border Color Right
       * @see https://tailwindcss.com/docs/border-color
       */
      'border-color-r': [{
        'border-r': [borderColor]
      }],

      /**
       * Border Color Bottom
       * @see https://tailwindcss.com/docs/border-color
       */
      'border-color-b': [{
        'border-b': [borderColor]
      }],

      /**
       * Border Color Left
       * @see https://tailwindcss.com/docs/border-color
       */
      'border-color-l': [{
        'border-l': [borderColor]
      }],

      /**
       * Divide Color
       * @see https://tailwindcss.com/docs/divide-color
       */
      'divide-color': [{
        divide: [borderColor]
      }],

      /**
       * Outline Style
       * @see https://tailwindcss.com/docs/outline-style
       */
      'outline-style': [{
        outline: [''].concat(getLineStyles(), ['hidden'])
      }],

      /**
       * Outline Offset
       * @see https://tailwindcss.com/docs/outline-offset
       */
      'outline-offset': [{
        'outline-offset': [isLength]
      }],

      /**
       * Outline Width
       * @see https://tailwindcss.com/docs/outline-width
       */
      'outline-w': [{
        outline: [isLength]
      }],

      /**
       * Outline Color
       * @see https://tailwindcss.com/docs/outline-color
       */
      'outline-color': [{
        outline: [colors]
      }],

      /**
       * Ring Width
       * @see https://tailwindcss.com/docs/ring-width
       */
      'ring-w': [{
        ring: getLengthWithEmpty()
      }],

      /**
       * Ring Width Inset
       * @see https://tailwindcss.com/docs/ring-width
       */
      'ring-w-inset': ['ring-inset'],

      /**
       * Ring Color
       * @see https://tailwindcss.com/docs/ring-color
       */
      'ring-color': [{
        ring: [colors]
      }],

      /**
       * Ring Opacity
       * @see https://tailwindcss.com/docs/ring-opacity
       */
      'ring-opacity': [{
        'ring-opacity': [opacity]
      }],

      /**
       * Ring Offset Width
       * @see https://tailwindcss.com/docs/ring-offset-width
       */
      'ring-offset-w': [{
        'ring-offset': [isLength]
      }],

      /**
       * Ring Offset Color
       * @see https://tailwindcss.com/docs/ring-offset-color
       */
      'ring-offset-color': [{
        'ring-offset': [colors]
      }],
      // Effects

      /**
       * Box Shadow
       * @see https://tailwindcss.com/docs/box-shadow
       */
      shadow: [{
        shadow: ['', 'inner', 'none', isTshirtSize, isArbitraryShadow]
      }],

      /**
       * Box Shadow Color
       * @see https://tailwindcss.com/docs/box-shadow-color
       */
      'shadow-color': [{
        shadow: [isAny]
      }],

      /**
       * Opacity
       * @see https://tailwindcss.com/docs/opacity
       */
      opacity: [{
        opacity: [opacity]
      }],

      /**
       * Mix Beldn Mode
       * @see https://tailwindcss.com/docs/mix-blend-mode
       */
      'mix-blend': [{
        'mix-blend': getBlendModes()
      }],

      /**
       * Background Blend Mode
       * @see https://tailwindcss.com/docs/background-blend-mode
       */
      'bg-blend': [{
        'bg-blend': getBlendModes()
      }],
      // Filters

      /**
       * Filter
       * @deprecated since Tailwind CSS v3.0.0
       * @see https://tailwindcss.com/docs/filter
       */
      filter: [{
        filter: ['', 'none']
      }],

      /**
       * Blur
       * @see https://tailwindcss.com/docs/blur
       */
      blur: [{
        blur: [blur]
      }],

      /**
       * Brightness
       * @see https://tailwindcss.com/docs/brightness
       */
      brightness: [{
        brightness: [brightness]
      }],

      /**
       * Contrast
       * @see https://tailwindcss.com/docs/contrast
       */
      contrast: [{
        contrast: [contrast]
      }],

      /**
       * Drop Shadow
       * @see https://tailwindcss.com/docs/drop-shadow
       */
      'drop-shadow': [{
        'drop-shadow': ['', 'none', isTshirtSize, isArbitraryValue]
      }],

      /**
       * Grayscale
       * @see https://tailwindcss.com/docs/grayscale
       */
      grayscale: [{
        grayscale: [grayscale]
      }],

      /**
       * Hue Rotate
       * @see https://tailwindcss.com/docs/hue-rotate
       */
      'hue-rotate': [{
        'hue-rotate': [hueRotate]
      }],

      /**
       * Invert
       * @see https://tailwindcss.com/docs/invert
       */
      invert: [{
        invert: [invert]
      }],

      /**
       * Saturate
       * @see https://tailwindcss.com/docs/saturate
       */
      saturate: [{
        saturate: [saturate]
      }],

      /**
       * Sepia
       * @see https://tailwindcss.com/docs/sepia
       */
      sepia: [{
        sepia: [sepia]
      }],

      /**
       * Backdrop Filter
       * @deprecated since Tailwind CSS v3.0.0
       * @see https://tailwindcss.com/docs/backdrop-filter
       */
      'backdrop-filter': [{
        'backdrop-filter': ['', 'none']
      }],

      /**
       * Backdrop Blur
       * @see https://tailwindcss.com/docs/backdrop-blur
       */
      'backdrop-blur': [{
        'backdrop-blur': [blur]
      }],

      /**
       * Backdrop Brightness
       * @see https://tailwindcss.com/docs/backdrop-brightness
       */
      'backdrop-brightness': [{
        'backdrop-brightness': [brightness]
      }],

      /**
       * Backdrop Contrast
       * @see https://tailwindcss.com/docs/backdrop-contrast
       */
      'backdrop-contrast': [{
        'backdrop-contrast': [contrast]
      }],

      /**
       * Backdrop Grayscale
       * @see https://tailwindcss.com/docs/backdrop-grayscale
       */
      'backdrop-grayscale': [{
        'backdrop-grayscale': [grayscale]
      }],

      /**
       * Backdrop Hue Rotate
       * @see https://tailwindcss.com/docs/backdrop-hue-rotate
       */
      'backdrop-hue-rotate': [{
        'backdrop-hue-rotate': [hueRotate]
      }],

      /**
       * Backdrop Invert
       * @see https://tailwindcss.com/docs/backdrop-invert
       */
      'backdrop-invert': [{
        'backdrop-invert': [invert]
      }],

      /**
       * Backdrop Opacity
       * @see https://tailwindcss.com/docs/backdrop-opacity
       */
      'backdrop-opacity': [{
        'backdrop-opacity': [opacity]
      }],

      /**
       * Backdrop Saturate
       * @see https://tailwindcss.com/docs/backdrop-saturate
       */
      'backdrop-saturate': [{
        'backdrop-saturate': [saturate]
      }],

      /**
       * Backdrop Sepia
       * @see https://tailwindcss.com/docs/backdrop-sepia
       */
      'backdrop-sepia': [{
        'backdrop-sepia': [sepia]
      }],
      // Tables

      /**
       * Border Collapse
       * @see https://tailwindcss.com/docs/border-collapse
       */
      'border-collapse': [{
        border: ['collapse', 'separate']
      }],

      /**
       * Table Layout
       * @see https://tailwindcss.com/docs/table-layout
       */
      'table-layout': [{
        table: ['auto', 'fixed']
      }],
      // Transitions and Animation

      /**
       * Tranisition Property
       * @see https://tailwindcss.com/docs/transition-property
       */
      transition: [{
        transition: ['none', 'all', '', 'colors', 'opacity', 'shadow', 'transform', isArbitraryValue]
      }],

      /**
       * Transition Duration
       * @see https://tailwindcss.com/docs/transition-duration
       */
      duration: [{
        duration: [isInteger]
      }],

      /**
       * Transition Timing Function
       * @see https://tailwindcss.com/docs/transition-timing-function
       */
      ease: [{
        ease: ['linear', 'in', 'out', 'in-out', isArbitraryValue]
      }],

      /**
       * Transition Delay
       * @see https://tailwindcss.com/docs/transition-delay
       */
      delay: [{
        delay: [isInteger]
      }],

      /**
       * Animation
       * @see https://tailwindcss.com/docs/animation
       */
      animate: [{
        animate: ['none', 'spin', 'ping', 'pulse', 'bounce', isArbitraryValue]
      }],
      // Transforms

      /**
       * Transform
       * @see https://tailwindcss.com/docs/transform
       */
      transform: [{
        transform: ['', 'gpu', 'none']
      }],

      /**
       * Scale
       * @see https://tailwindcss.com/docs/scale
       */
      scale: [{
        scale: [scale]
      }],

      /**
       * Scale X
       * @see https://tailwindcss.com/docs/scale
       */
      'scale-x': [{
        'scale-x': [scale]
      }],

      /**
       * Scale Y
       * @see https://tailwindcss.com/docs/scale
       */
      'scale-y': [{
        'scale-y': [scale]
      }],

      /**
       * Rotate
       * @see https://tailwindcss.com/docs/rotate
       */
      rotate: [{
        rotate: [isInteger, isArbitraryValue]
      }],

      /**
       * Translate X
       * @see https://tailwindcss.com/docs/translate
       */
      'translate-x': [{
        'translate-x': [translate]
      }],

      /**
       * Translate Y
       * @see https://tailwindcss.com/docs/translate
       */
      'translate-y': [{
        'translate-y': [translate]
      }],

      /**
       * Skew X
       * @see https://tailwindcss.com/docs/skew
       */
      'skew-x': [{
        'skew-x': [skew]
      }],

      /**
       * Skew Y
       * @see https://tailwindcss.com/docs/skew
       */
      'skew-y': [{
        'skew-y': [skew]
      }],

      /**
       * Transform Origin
       * @see https://tailwindcss.com/docs/transform-origin
       */
      'transform-origin': [{
        origin: ['center', 'top', 'top-right', 'right', 'bottom-right', 'bottom', 'bottom-left', 'left', 'top-left', isArbitraryValue]
      }],
      // Interactivity

      /**
       * Accent Color
       * @see https://tailwindcss.com/docs/accent-color
       */
      accent: [{
        accent: ['auto', colors]
      }],

      /**
       * Appearance
       * @see https://tailwindcss.com/docs/appearance
       */
      appearance: ['appearance-none'],

      /**
       * Cursor
       * @see https://tailwindcss.com/docs/cursor
       */
      cursor: [{
        cursor: ['auto', 'default', 'pointer', 'wait', 'text', 'move', 'help', 'not-allowed', 'none', 'context-menu', 'progress', 'cell', 'crosshair', 'vertical-text', 'alias', 'copy', 'no-drop', 'grab', 'grabbing', 'all-scroll', 'col-resize', 'row-resize', 'n-resize', 'e-resize', 's-resize', 'w-resize', 'ne-resize', 'nw-resize', 'se-resize', 'sw-resize', 'ew-resize', 'ns-resize', 'nesw-resize', 'nwse-resize', 'zoom-in', 'zoom-out', isArbitraryValue]
      }],

      /**
       * Caret Color
       * @see https://tailwindcss.com/docs/just-in-time-mode#caret-color-utilities
       */
      'caret-color': [{
        caret: [colors]
      }],

      /**
       * Pointer Events
       * @see https://tailwindcss.com/docs/pointer-events
       */
      'pointer-events': [{
        'pointer-events': ['none', 'auto']
      }],

      /**
       * Resize
       * @see https://tailwindcss.com/docs/resize
       */
      resize: [{
        resize: ['none', 'y', 'x', '']
      }],

      /**
       * Scroll Behavior
       * @see https://tailwindcss.com/docs/scroll-behavior
       */
      'scroll-behavior': [{
        scroll: ['auto', 'smooth']
      }],

      /**
       * Scroll Margin
       * @see https://tailwindcss.com/docs/scroll-margin
       */
      'scroll-m': [{
        'scroll-m': [spacing]
      }],

      /**
       * Scroll Margin X
       * @see https://tailwindcss.com/docs/scroll-margin
       */
      'scroll-mx': [{
        'scroll-mx': [spacing]
      }],

      /**
       * Scroll Margin Y
       * @see https://tailwindcss.com/docs/scroll-margin
       */
      'scroll-my': [{
        'scroll-my': [spacing]
      }],

      /**
       * Scroll Margin Top
       * @see https://tailwindcss.com/docs/scroll-margin
       */
      'scroll-mt': [{
        'scroll-mt': [spacing]
      }],

      /**
       * Scroll Margin Right
       * @see https://tailwindcss.com/docs/scroll-margin
       */
      'scroll-mr': [{
        'scroll-mr': [spacing]
      }],

      /**
       * Scroll Margin Bottom
       * @see https://tailwindcss.com/docs/scroll-margin
       */
      'scroll-mb': [{
        'scroll-mb': [spacing]
      }],

      /**
       * Scroll Margin Left
       * @see https://tailwindcss.com/docs/scroll-margin
       */
      'scroll-ml': [{
        'scroll-ml': [spacing]
      }],

      /**
       * Scroll Padding
       * @see https://tailwindcss.com/docs/scroll-padding
       */
      'scroll-p': [{
        'scroll-p': [spacing]
      }],

      /**
       * Scroll Padding X
       * @see https://tailwindcss.com/docs/scroll-padding
       */
      'scroll-px': [{
        'scroll-px': [spacing]
      }],

      /**
       * Scroll Padding Y
       * @see https://tailwindcss.com/docs/scroll-padding
       */
      'scroll-py': [{
        'scroll-py': [spacing]
      }],

      /**
       * Scroll Padding Top
       * @see https://tailwindcss.com/docs/scroll-padding
       */
      'scroll-pt': [{
        'scroll-pt': [spacing]
      }],

      /**
       * Scroll Padding Right
       * @see https://tailwindcss.com/docs/scroll-padding
       */
      'scroll-pr': [{
        'scroll-pr': [spacing]
      }],

      /**
       * Scroll Padding Bottom
       * @see https://tailwindcss.com/docs/scroll-padding
       */
      'scroll-pb': [{
        'scroll-pb': [spacing]
      }],

      /**
       * Scroll Padding Left
       * @see https://tailwindcss.com/docs/scroll-padding
       */
      'scroll-pl': [{
        'scroll-pl': [spacing]
      }],

      /**
       * Scroll Snap Align
       * @see https://tailwindcss.com/docs/scroll-snap-align
       */
      'snap-align': [{
        snap: ['start', 'end', 'center', 'align-none']
      }],

      /**
       * Scroll Snap Stop
       * @see https://tailwindcss.com/docs/scroll-snap-stop
       */
      'snap-stop': [{
        snap: ['normal', 'always']
      }],

      /**
       * Scroll Snap Type
       * @see https://tailwindcss.com/docs/scroll-snap-type
       */
      'snap-type': [{
        snap: ['none', 'x', 'y', 'both']
      }],

      /**
       * Scroll Snap Type Strictness
       * @see https://tailwindcss.com/docs/scroll-snap-type
       */
      'snap-strictness': [{
        snap: ['mandatory', 'proximity']
      }],

      /**
       * Touch Action
       * @see https://tailwindcss.com/docs/touch-action
       */
      touch: [{
        touch: ['auto', 'none', 'pinch-zoom', 'manipulation', {
          pan: ['x', 'left', 'right', 'y', 'up', 'down']
        }]
      }],

      /**
       * User Select
       * @see https://tailwindcss.com/docs/user-select
       */
      select: [{
        select: ['none', 'text', 'all', 'auto']
      }],

      /**
       * Will Change
       * @see https://tailwindcss.com/docs/will-change
       */
      'will-change': [{
        'will-change': ['auto', 'scroll', 'contents', 'transform', isArbitraryValue]
      }],
      // SVG

      /**
       * Fill
       * @see https://tailwindcss.com/docs/fill
       */
      fill: [{
        fill: [colors]
      }],

      /**
       * Stroke Width
       * @see https://tailwindcss.com/docs/stroke-width
       */
      'stroke-w': [{
        stroke: [isLength]
      }],

      /**
       * Stroke
       * @see https://tailwindcss.com/docs/stroke
       */
      stroke: [{
        stroke: [colors]
      }],
      // Accessibility

      /**
       * Screen Readers
       * @see https://tailwindcss.com/docs/screen-readers
       */
      sr: ['sr-only', 'not-sr-only']
    },
    conflictingClassGroups: {
      overflow: ['overflow-x', 'overflow-y'],
      overscroll: ['overscroll-x', 'overscroll-y'],
      inset: ['inset-x', 'inset-y', 'top', 'right', 'bottom', 'left'],
      'inset-x': ['right', 'left'],
      'inset-y': ['top', 'bottom'],
      flex: ['basis', 'grow', 'shrink'],
      'col-start-end': ['col-start', 'col-end'],
      'row-start-end': ['row-start', 'row-end'],
      gap: ['gap-x', 'gap-y'],
      p: ['px', 'py', 'pt', 'pr', 'pb', 'pl'],
      px: ['pr', 'pl'],
      py: ['pt', 'pb'],
      m: ['mx', 'my', 'mt', 'mr', 'mb', 'ml'],
      mx: ['mr', 'ml'],
      my: ['mt', 'mb'],
      'font-size': ['leading'],
      'fvn-normal': ['fvn-ordinal', 'fvn-slashed-zero', 'fvn-figure', 'fvn-spacing', 'fvn-fraction'],
      'fvn-ordinal': ['fvn-normal'],
      'fvn-slashed-zero': ['fvn-normal'],
      'fvn-figure': ['fvn-normal'],
      'fvn-spacing': ['fvn-normal'],
      'fvn-fraction': ['fvn-normal'],
      rounded: ['rounded-t', 'rounded-r', 'rounded-b', 'rounded-l', 'rounded-tl', 'rounded-tr', 'rounded-br', 'rounded-bl'],
      'rounded-t': ['rounded-tl', 'rounded-tr'],
      'rounded-r': ['rounded-tr', 'rounded-br'],
      'rounded-b': ['rounded-br', 'rounded-bl'],
      'rounded-l': ['rounded-tl', 'rounded-bl'],
      'border-w': ['border-w-t', 'border-w-r', 'border-w-b', 'border-w-l'],
      'border-w-x': ['border-w-r', 'border-w-l'],
      'border-w-y': ['border-w-t', 'border-w-b'],
      'border-color': ['border-color-t', 'border-color-r', 'border-color-b', 'border-color-l'],
      'border-color-x': ['border-color-r', 'border-color-l'],
      'border-color-y': ['border-color-t', 'border-color-b'],
      'scroll-m': ['scroll-mx', 'scroll-my', 'scroll-mt', 'scroll-mr', 'scroll-mb', 'scroll-ml'],
      'scroll-mx': ['scroll-mr', 'scroll-ml'],
      'scroll-my': ['scroll-mt', 'scroll-mb'],
      'scroll-p': ['scroll-px', 'scroll-py', 'scroll-pt', 'scroll-pr', 'scroll-pb', 'scroll-pl'],
      'scroll-px': ['scroll-pr', 'scroll-pl'],
      'scroll-py': ['scroll-pt', 'scroll-pb']
    }
  };
}

var twMerge = /*#__PURE__*/createTailwindMerge(getDefaultConfig);

const _excluded$1e = ["children", "href", "shape", "size", "variant", "color", "startIcon", "endIcon", "fullWidth", "responsive", "animation", "loading", "active", "disabled", "dataTheme", "className", "style"];
const Button = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    children,
    href,
    shape,
    size,
    variant,
    color,
    startIcon,
    endIcon,
    fullWidth,
    responsive,
    animation = true,
    loading,
    active,
    disabled,
    dataTheme,
    className,
    style
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$1e);

  const classes = twMerge('btn', className, clsx((startIcon && !loading || endIcon) && 'gap-2', {
    [`btn-${size}`]: size,
    [`btn-${shape}`]: shape,
    [`btn-${variant}`]: variant,
    [`btn-${color}`]: color,
    'btn-block': fullWidth,
    'btn-xs md:btn-sm lg:btn-md xl:btn-lg': responsive,
    'no-animation': !animation,
    'btn-active': active,
    'btn-disabled': disabled,
    loading: loading
  }));

  if (href) {
    return jsxs("a", {
      className: classes,
      style: style,
      href: href,
      children: [startIcon && startIcon, children, endIcon && endIcon]
    });
  } else {
    return jsxs("button", _extends$1({}, props, {
      ref: ref,
      "data-theme": dataTheme,
      className: classes,
      style: style,
      disabled: disabled,
      children: [startIcon && !loading && startIcon, children, endIcon && endIcon]
    }));
  }
});
Button.displayName = 'Button';

const _excluded$1d = ["dataTheme", "className"];

const DropdownMenu = _ref => {
  let {
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$1d);

  const classes = twMerge('dropdown-content menu p-2 shadow bg-base-100 rounded-box', className);
  return jsx("ul", _extends$1({}, props, {
    tabIndex: 0,
    "data-theme": dataTheme,
    className: classes
  }));
};

const _excluded$1c = ["className"];
const DropdownItem = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$1c);

  return jsx("li", {
    className: className,
    children: jsx("a", _extends$1({
      ref: ref
    }, props))
  });
});

const _excluded$1b = ["children", "color", "size", "button", "dataTheme", "className", "disabled"];

const DropdownToggle = _ref => {
  let {
    children,
    color,
    size,
    button = true,
    dataTheme,
    className,
    disabled
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$1b);

  return jsx("label", _extends$1({
    tabIndex: 0,
    className: className
  }, props, {
    children: button ? jsx(Button, {
      dataTheme: dataTheme,
      color: color,
      size: size,
      disabled: disabled,
      children: children
    }) : children
  }));
};

const _excluded$1a = ["children", "item", "horizontal", "vertical", "hover", "open", "dataTheme"];
const Dropdown = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    item,
    horizontal,
    vertical,
    hover,
    open,
    dataTheme
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$1a);

  const classes = twMerge('dropdown', clsx({
    [`dropdown-${horizontal}`]: horizontal,
    [`dropdown-${vertical}`]: vertical,
    'dropdown-hover': hover,
    'dropdown-open': open
  }));
  return jsxs("div", _extends$1({
    role: "listbox"
  }, props, {
    ref: ref,
    "data-theme": dataTheme,
    className: classes,
    children: [jsx("label", {
      tabIndex: 0,
      children: children
    }), jsx("ul", {
      className: "dropdown-content",
      children: item
    })]
  }));
});
var Dropdown$1 = Object.assign(Dropdown, {
  Toggle: DropdownToggle,
  Menu: DropdownMenu,
  Item: DropdownItem
});

const _excluded$19 = ["children", "className"];
const ModalActions = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$19);

  const classes = twMerge('modal-action', className);
  return jsx("div", _extends$1({}, props, {
    className: classes,
    ref: ref,
    children: children
  }));
});
ModalActions.displayName = 'ModalActions';

const _excluded$18 = ["children", "className"];
const ModalBody = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$18);

  return jsx("div", _extends$1({}, props, {
    className: className,
    ref: ref,
    children: children
  }));
});
ModalBody.displayName = 'ModalBody';

const _excluded$17 = ["children", "className"];
const ModalHeader = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$17);

  const classes = twMerge('w-full mb-8 text-xl', className);
  return jsx("div", _extends$1({}, props, {
    className: classes,
    ref: ref,
    children: children
  }));
});
ModalHeader.displayName = 'ModalHeader';

const _excluded$16 = ["children", "open", "responsive", "onClickBackdrop", "dataTheme", "className"];
const Modal = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    children,
    open,
    responsive,
    onClickBackdrop,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$16);

  const containerClasses = twMerge('modal', clsx({
    'modal-open': open,
    'modal-bottom sm:modal-middle': responsive
  }));
  const bodyClasses = twMerge('modal-box', className);
  return jsx("div", {
    "aria-label": "Modal",
    "aria-hidden": !open,
    "aria-modal": open,
    "data-theme": dataTheme,
    className: containerClasses,
    onClick: e => {
      e.stopPropagation();

      if (e.target === e.currentTarget) {
        e.stopPropagation();

        if (onClickBackdrop) {
          onClickBackdrop();
        }
      }
    },
    children: jsx("div", _extends$1({}, props, {
      "data-theme": dataTheme,
      className: bodyClasses,
      ref: ref,
      children: children
    }))
  });
});
Modal.displayName = 'Modal';
var Modal$1 = Object.assign(Modal, {
  Header: ModalHeader,
  Body: ModalBody,
  Actions: ModalActions
});

const isReactFragment = node => {
  if (!node) return false;

  if (node != null && node.type) {
    return (node == null ? void 0 : node.type) === React.Fragment;
  }

  return node === React.Fragment;
}; // Return true if a react element is valid and not a fragment
// If a valid element is passed, add the props

const wrapWithElementIfInvalid = ({
  node,
  wrapper,
  props: _props = {}
}) => {
  if (!node) {
    return /*#__PURE__*/React.cloneElement(wrapper, _props);
  } else if (! /*#__PURE__*/React.isValidElement(node)) {
    return /*#__PURE__*/React.cloneElement(wrapper, _props, node);
  } else if (isReactFragment(node)) {
    var _node$props;

    return /*#__PURE__*/React.cloneElement(wrapper, _extends$1({}, _props, {
      className: twMerge((_node$props = node.props) == null ? void 0 : _node$props.className, _props == null ? void 0 : _props.className)
    }), node.props.children);
  } else {
    var _node$props2;

    return /*#__PURE__*/React.cloneElement(node, _extends$1({}, _props, {
      className: twMerge((_node$props2 = node.props) == null ? void 0 : _node$props2.className, _props == null ? void 0 : _props.className)
    }));
  }
}; // Returns true if there is a single, string child element

const isSingleStringChild = children => {
  return children && React.Children.count(children) === 1 && /*#__PURE__*/React.isValidElement(children) && typeof children.props.children === 'string';
};

const _excluded$15 = ["onElement", "offElement", "active", "rotate", "flip", "dataTheme", "className"];
const Swap = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    onElement,
    offElement,
    active,
    rotate,
    flip,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$15);

  const classes = twMerge('swap', className, clsx({
    'swap-active': active,
    'swap-rotate': rotate,
    'swap-flip': flip
  })); // These next two pieces allow classname to be added to valid elements, or wrap invalid elements with a div and the classname

  const onEl = wrapWithElementIfInvalid({
    node: onElement,
    wrapper: jsx("div", {}),
    props: {
      className: 'swap-on'
    }
  });
  const offEl = wrapWithElementIfInvalid({
    node: offElement,
    wrapper: jsx("div", {}),
    props: {
      className: 'swap-off'
    }
  });
  return jsxs("label", _extends$1({}, props, {
    "data-theme": dataTheme,
    className: classes,
    ref: ref,
    children: [jsx("input", {
      type: "checkbox"
    }), onEl, offEl]
  }));
});

const _excluded$14 = ["children", "icon", "status", "dataTheme", "className", "innerClassName"];
const Alert = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    children,
    icon,
    status,
    dataTheme,
    className,
    innerClassName
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$14);

  const classes = twMerge('alert', className, clsx({
    [`alert-${status}`]: status
  }));
  return jsx("div", _extends$1({
    role: "alert"
  }, props, {
    ref: ref,
    "data-theme": dataTheme,
    className: classes,
    children: jsxs("div", {
      className: twMerge('flex-1', innerClassName),
      children: [icon, children]
    })
  }));
});
Alert.displayName = 'Alert';

const _excluded$13 = ["children", "className"];
const AvatarGroup = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$13);

  const classes = clsx('avatar-group', '-space-x-6', className);
  return jsx("div", _extends$1({
    "aria-label": `Group of ${children.length} avatar photos`
  }, props, {
    className: classes,
    ref: ref,
    children: children
  }));
});

const _excluded$12 = ["size", "src", "letters", "shape", "color", "border", "borderColor", "online", "offline", "dataTheme", "className", "children"];
const Avatar = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    size = 'md',
    src,
    letters,
    shape,
    color,
    border,
    borderColor,
    online,
    offline,
    dataTheme,
    className,
    children
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$12);

  const containerClasses = twMerge('avatar', className, clsx({
    online: online,
    offline: offline,
    placeholder: !src
  }));
  const imgClasses = clsx({
    'ring ring-offset-base-100 ring-offset-2': border,
    [`ring-${borderColor}`]: borderColor,
    'rounded-btn': shape === 'square',
    'rounded-full': shape === 'circle',
    'w-32 h-32': size === 'lg',
    'w-24 h-24': size === 'md',
    'w-14 h-14': size === 'sm',
    'w-10 h-10': size === 'xs'
  });
  const placeholderClasses = clsx({
    'bg-neutral-focus': !color,
    'text-neutral-content': !color,
    [`bg-${color}`]: color,
    [`text-${color}-content`]: color,
    'ring ring-offset-base-100 ring-offset-2': border,
    [`ring-${borderColor}`]: borderColor,
    'rounded-btn': shape === 'square',
    'rounded-full': shape === 'circle',
    'w-32 h-32 text-3xl': size === 'lg',
    'w-24 h-24 text-xl': size === 'md',
    'w-14 h-14': size === 'sm',
    'w-10 h-10': size === 'xs'
  });
  const customImgDimension = typeof size === 'number' ? {
    width: size,
    height: size
  } : {};

  const renderAvatarContents = () => {
    // Base case, if src is provided, render img
    if (src) {
      return jsx("div", {
        className: imgClasses,
        style: customImgDimension,
        children: jsx("img", {
          src: src
        })
      });
    } // Render a text avatar if letters are provided, or a single child that is a string
    else if (letters || isSingleStringChild(children)) {
      return jsx("div", {
        className: placeholderClasses,
        style: customImgDimension,
        children: jsx("span", {
          children: letters ? letters : children
        })
      });
    } // Render if a single, not string child was provided (allows for SVGs) and merges classes and styles
    else if (React.Children.count(children) === 1) {
      const firstChild = React.Children.only(children);
      return /*#__PURE__*/React.cloneElement(firstChild, {
        className: twMerge(imgClasses, firstChild.props.className),
        style: _extends$1({}, customImgDimension, firstChild.props.style)
      });
    } // Render a wrapping div around all children if there is more than one child.
    else {
      return jsx("div", {
        className: imgClasses,
        style: customImgDimension,
        children: children
      });
    }
  };

  return jsx("div", _extends$1({
    "aria-label": "Avatar photo"
  }, props, {
    "data-theme": dataTheme,
    className: containerClasses,
    ref: ref,
    children: renderAvatarContents()
  }));
});
var Avatar$1 = Object.assign(Avatar, {
  Group: AvatarGroup
});

const _excluded$11 = ["children", "variant", "size", "color", "responsive", "dataTheme", "className"];
const Badge = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    children,
    variant,
    size,
    color,
    responsive,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$11);

  const classes = twMerge('badge', className, clsx({
    [`badge-${size}`]: size,
    [`badge-${variant}`]: variant,
    [`badge-${color}`]: color,
    'badge-xs md:badge-sm lg:badge-md xl:badge-lg': responsive
  }));
  return jsx("div", _extends$1({
    "aria-label": "Badge"
  }, props, {
    "data-theme": dataTheme,
    className: classes,
    ref: ref,
    children: children
  }));
});
Badge.displayName = 'Badge';

const _excluded$10 = ["className"];
const CardActions = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$10);

  return jsx("div", _extends$1({}, props, {
    className: twMerge('card-actions', className),
    ref: ref
  }));
});

const _excluded$$ = ["className"];
const CardBody = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$$);

  return jsx("div", _extends$1({}, props, {
    className: twMerge('card-body', className),
    ref: ref
  }));
});

const _excluded$_ = ["className", "tag"];
const CardTitle = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    className,
    tag = 'div'
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$_);

  const Tag = tag;
  return jsx(Tag, _extends$1({}, props, {
    className: twMerge('card-title', className),
    ref: ref
  }));
});

const CardImage = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let props = _extends$1({}, _ref);

  return jsx("figure", {
    ref: ref,
    children: jsx("img", _extends$1({}, props))
  });
});

const _excluded$Z = ["bordered", "imageFull", "normal", "compact", "side", "className"];
const DYNAMIC_MODIFIERS$1 = {
  compact: {
    true: 'card-compact',
    xs: 'xs:card-compact',
    sm: 'sm:card-compact',
    md: 'md:card-compact',
    lg: 'lg:card-compact'
  },
  normal: {
    true: 'card-normal',
    xs: 'xs:card-normal',
    sm: 'sm:card-normal',
    md: 'md:card-normal',
    lg: 'lg:card-normal'
  },
  side: {
    true: 'card-side',
    xs: 'xs:card-side',
    sm: 'sm:card-side',
    md: 'md:card-side',
    lg: 'lg:card-side'
  }
};
const Card = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    bordered = true,
    imageFull,
    normal,
    compact,
    side,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$Z);

  const classes = twMerge('card', className, clsx({
    'card-bordered': bordered,
    'image-full': imageFull,
    [compact && DYNAMIC_MODIFIERS$1.compact[compact.toString()] || '']: compact,
    [normal && DYNAMIC_MODIFIERS$1.normal[normal.toString()] || '']: normal,
    [side && DYNAMIC_MODIFIERS$1.side[side.toString()] || '']: side
  }));
  return jsx("div", _extends$1({
    "aria-label": "Card"
  }, props, {
    className: classes,
    ref: ref
  }));
});
var Card$1 = Object.assign(Card, {
  Actions: CardActions,
  Body: CardBody,
  Title: CardTitle,
  Image: CardImage
});

const _excluded$Y = ["children", "innerRef", "src", "alt", "index", "width", "hasButtons", "buttonStyle", "onPrev", "onNext", "className"];

const CarouselItem = _ref => {
  let {
    children,
    innerRef,
    src,
    alt,
    index = 0,
    width,
    hasButtons,
    buttonStyle,
    onPrev,
    onNext,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$Y);

  const classes = twMerge('carousel-item relative', className, clsx({
    'w-full': width === 'full',
    'w-1/2': width === 'half',
    'h-full': true
  }));
  const imageClasses = clsx({
    'w-full': width === 'full'
  });

  const renderButtons = () => {
    if (buttonStyle != null) {
      return jsxs(Fragment, {
        children: [/*#__PURE__*/cloneElement(buttonStyle('â®'), {
          onClick: onPrev
        }), /*#__PURE__*/cloneElement(buttonStyle('â¯'), {
          onClick: onNext
        })]
      });
    }

    return jsxs(Fragment, {
      children: [jsx(Button, {
        onClick: onPrev,
        shape: "circle",
        children: "\u276E"
      }), jsx(Button, {
        onClick: onNext,
        shape: "circle",
        children: "\u276F"
      })]
    });
  };

  return jsxs("div", _extends$1({}, props, {
    id: `item${index}`,
    ref: innerRef,
    className: classes,
    children: [src ? jsx("img", {
      src: src,
      alt: alt,
      className: imageClasses
    }) : children, hasButtons && jsx("div", {
      className: "absolute flex justify-between transform -translate-y-1/2 left-5 right-5 top-1/2",
      children: renderButtons()
    })]
  }));
};

const _excluded$X = ["children", "display", "snap", "vertical", "width", "buttonStyle", "dataTheme", "className"];
const Carousel = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    children,
    display = 'slider',
    snap,
    vertical,
    width,
    buttonStyle,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$X);

  const classes = twMerge('carousel', className, clsx({
    [`carousel-${snap}`]: snap,
    'carousel-vertical': vertical,
    'w-full': display !== 'slider'
  }));
  const [itemRefs, setItemRefs] = useState([]);
  const [activeIndex, setActiveIndex] = useState(0);
  useEffect(() => {
    const newRefs = [];
    children.map(_ => {
      newRefs.push( /*#__PURE__*/createRef());
    });
    setItemRefs(newRefs);
  }, [children]);

  const scrollToIndex = index => {
    var _itemRefs$index$curre;

    (_itemRefs$index$curre = itemRefs[index].current) == null ? void 0 : _itemRefs$index$curre.scrollIntoView({
      behavior: 'smooth',
      block: 'nearest',
      inline: snap
    });
    setActiveIndex(index);
  };

  return jsxs(Fragment, {
    children: [jsx("div", _extends$1({
      role: "listbox",
      "aria-label": "Image carousel"
    }, props, {
      ref: ref,
      "data-theme": dataTheme,
      className: classes,
      children: children.map((child, i) => {
        return /*#__PURE__*/cloneElement(child, _extends$1({
          innerRef: itemRefs[i],
          index: i + 1,
          children: child.props.children,
          src: child.props.src,
          alt: child.props.alt,
          width: display !== 'slider' ? 'full' : width,
          hasButtons: display === 'sequential',
          buttonStyle,
          onPrev: () => scrollToIndex(i - 1 < 0 ? children.length - 1 : i - 1),
          onNext: () => scrollToIndex(i + 1 > children.length - 1 ? 0 : i + 1)
        }, child.props));
      })
    })), display === 'numbered' && jsx("div", {
      className: "flex justify-center w-full py-2 gap-2",
      children: children.map((_, i) => {
        if (buttonStyle != null) {
          return /*#__PURE__*/cloneElement(buttonStyle((i + 1).toString()), {
            key: i,
            onClick: () => scrollToIndex(i)
          });
        }

        return jsx(Button, {
          active: i === activeIndex,
          onClick: () => scrollToIndex(i),
          children: i + 1
        }, i);
      })
    })]
  });
});
var Carousel$1 = Object.assign(Carousel, {
  Item: CarouselItem
});

const _excluded$W = ["children", "className"];

const CollapseTitle = _ref => {
  let {
    children,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$W);

  const classes = twMerge('collapse-title', className);
  return jsx("div", _extends$1({}, props, {
    className: classes,
    children: children
  }));
};

const _excluded$V = ["children", "className"];

const CollapseContent = _ref => {
  let {
    children,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$V);

  const classes = twMerge('collapse-content', className);
  return jsx("div", _extends$1({}, props, {
    className: classes,
    children: children
  }));
};

const _excluded$U = ["children", "checkbox", "icon", "open", "dataTheme", "className", "onOpen", "onClose", "onToggle"];
const Collapse = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    checkbox,
    icon,
    open,
    dataTheme,
    className,
    onOpen,
    onClose,
    onToggle
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$U);

  const classes = twMerge('collapse', className, clsx({
    [`collapse-${icon}`]: icon,
    'collapse-open': open === true,
    'collapse-close': open === false
  }));
  const checkboxRef = useRef(null); // Handle events for checkbox changes

  const handleCheckboxChange = () => {
    var _checkboxRef$current, _checkboxRef$current2;

    if (onToggle) {
      onToggle();
    }

    if (onOpen && (_checkboxRef$current = checkboxRef.current) != null && _checkboxRef$current.checked) {
      onOpen();
    } else if (onClose && !((_checkboxRef$current2 = checkboxRef.current) != null && _checkboxRef$current2.checked)) {
      onClose();
    }
  }; // Handle blur events, specifically handling open/close for non checkbox types


  const handleBlur = event => {
    if (!checkbox && onToggle) onToggle();
    if (!checkbox && onClose) onClose();
    if (props.onBlur) props.onBlur(event);
  }; // Handle focus events, specifically handling open/close for non checkbox types


  const handleFocus = event => {
    if (!checkbox && onToggle) onToggle();
    if (!checkbox && onOpen) onOpen();
    if (props.onFocus) props.onFocus(event);
  };

  return jsxs("div", _extends$1({
    "aria-expanded": open
  }, props, {
    ref: ref,
    tabIndex: 0,
    "data-theme": dataTheme,
    className: classes,
    onBlur: handleBlur,
    onFocus: handleFocus,
    children: [checkbox && jsx("input", {
      type: "checkbox",
      className: "peer",
      ref: checkboxRef,
      onChange: handleCheckboxChange
    }), children]
  }));
});
var Collapse$1 = Object.assign(Collapse, {
  Title: CollapseTitle,
  Content: CollapseContent
});

const _excluded$T = ["value", "dataTheme", "className"];

const Countdown = _ref => {
  let {
    value,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$T);

  const classes = twMerge('countdown', className);
  const displayedValue = Math.min(99, Math.max(0, value));
  const countdownStyle = {
    '--value': displayedValue
  };
  return jsx("span", _extends$1({
    role: "timer"
  }, props, {
    "data-theme": dataTheme,
    className: classes,
    children: jsx("span", {
      style: countdownStyle
    })
  }));
};

const _excluded$S = ["children", "size", "dataTheme", "className"];
const Kbd = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    children,
    size,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$S);

  const classes = twMerge('kbd', className, clsx({
    [`kbd-${size}`]: size
  }));
  return jsx("kbd", _extends$1({}, props, {
    "data-theme": dataTheme,
    className: classes,
    ref: ref,
    children: children
  }));
});

const _excluded$R = ["color", "dataTheme", "className"];
const Progress = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    color,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$R);

  const classes = twMerge('progress', className, clsx({
    [`progress-${color}`]: color
  }));
  return jsx("progress", _extends$1({}, props, {
    ref: ref,
    "data-theme": dataTheme,
    className: classes
  }));
});
Progress.displayName = 'Progress';

const _excluded$Q = ["value", "size", "thickness", "color", "dataTheme", "className", "children"];
const RadialProgress = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    value,
    size = '4rem',
    thickness = '4px',
    color,
    dataTheme,
    className,
    children
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$Q);

  const classes = twMerge('radial-progress', className, clsx({
    [`text-${color}`]: color
  }));
  const displayedValue = Math.min(100, Math.max(0, value));
  const progressStyle = {
    '--value': displayedValue,
    '--size': size,
    '--thickness': thickness
  };
  return jsx("div", _extends$1({
    role: "progressbar",
    "aria-valuenow": displayedValue,
    "aria-valuemin": 0,
    "aria-valuemax": 100
  }, props, {
    ref: ref,
    "data-theme": dataTheme,
    className: classes,
    style: progressStyle,
    children: children
  }));
});
RadialProgress.displayName = 'RadialProgress';

const _excluded$P = ["variant", "className"];
const StatItem = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    variant,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$P);

  const classes = twMerge(className, clsx({
    [`stat-${variant}`]: variant
  }));
  return jsx("div", _extends$1({}, props, {
    className: classes,
    ref: ref
  }));
});

const _excluded$O = ["dataTheme", "className"];
const Stat = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$O);

  const classes = twMerge('stat', className);
  return jsx("div", _extends$1({}, props, {
    "data-theme": dataTheme,
    className: classes,
    ref: ref
  }));
});
var Stat$1 = Object.assign(Stat, {
  Item: StatItem
});

const _excluded$N = ["horizontal", "vertical", "dataTheme", "className", "children"];
const Stats = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    horizontal,
    vertical,
    dataTheme,
    className,
    children
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$N);

  const classes = twMerge('stats', className, clsx({
    'stats-horizontal': horizontal,
    'stats-vertical': vertical
  }));
  return jsx("div", _extends$1({}, props, {
    ref: ref,
    "data-theme": dataTheme,
    className: classes,
    children: children
  }));
});
var Stats$1 = Object.assign(Stats, {
  Stat: Stat$1
});

const _excluded$M = ["children"];
const TableHead = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$M);

  return jsx("thead", _extends$1({}, props, {
    ref: ref,
    children: jsx("tr", {
      children: children == null ? void 0 : children.map(child => {
        return jsx("th", {
          children: child
        });
      })
    })
  }));
});

const _excluded$L = ["children"];
const TableBody = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$L);

  return jsx("tbody", _extends$1({}, props, {
    ref: ref,
    children: children
  }));
});

const _excluded$K = ["children", "active", "hover", "className"];
const TableRow = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    active,
    hover,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$K);

  const classes = twMerge(className, clsx({
    active: active,
    hover: hover
  }));
  return jsx("tr", _extends$1({}, props, {
    className: classes,
    ref: ref,
    children: children == null ? void 0 : children.map((child, i) => i < 1 ? jsx("th", {
      children: child
    }) : jsx("td", {
      children: child
    }))
  }));
});

const _excluded$J = ["children"];
const TableFooter = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$J);

  return jsx("tfoot", _extends$1({}, props, {
    ref: ref,
    children: jsx("tr", {
      children: children == null ? void 0 : children.map(child => {
        return jsx("th", {
          children: child
        });
      })
    })
  }));
});

const _excluded$I = ["children", "compact", "zebra", "dataTheme", "className"];
const Table = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    compact,
    zebra,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$I);

  const classes = twMerge('table', className, clsx({
    'table-zebra': zebra,
    'table-compact': compact
  }));
  return jsx("table", _extends$1({}, props, {
    "data-theme": dataTheme,
    className: classes,
    ref: ref,
    children: children
  }));
});
var Table$1 = Object.assign(Table, {
  Head: TableHead,
  Body: TableBody,
  Row: TableRow,
  Footer: TableFooter
});

const _excluded$H = ["message", "children", "open", "color", "position", "dataTheme", "className"];
const Tooltip = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    message,
    children,
    open,
    color,
    position,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$H);

  const classes = twMerge('tooltip', className, clsx({
    'tooltip-open': open,
    [`tooltip-${color}`]: color,
    [`tooltip-${position}`]: position
  }));
  return jsx("div", _extends$1({
    role: "tooltip"
  }, props, {
    ref: ref,
    "data-theme": dataTheme,
    "data-tip": message,
    className: classes,
    children: children
  }));
});
Tooltip.displayName = 'Tooltip';

const _excluded$G = ["children", "title", "dataTheme", "className"];
const Label = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    title,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$G);

  const classes = twMerge('label', className);
  return jsxs("label", _extends$1({}, props, {
    className: classes,
    children: [jsx("span", {
      className: "label-text cursor-pointer",
      ref: ref,
      children: title
    }), children]
  }));
});

const _excluded$F = ["children", "dataTheme", "className"];
const Form = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$F);

  const classes = twMerge('form-control', className);
  return jsx("div", _extends$1({
    role: "form"
  }, props, {
    "data-theme": dataTheme,
    className: classes,
    ref: ref,
    children: children
  }));
});
var Form$1 = Object.assign(Form, {
  Label
});

const _excluded$E = ["checked", "defaultChecked", "color", "size", "indeterminate", "dataTheme", "className"];
const Checkbox = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    checked,
    defaultChecked = false,
    color,
    size,
    indeterminate,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$E);

  const classes = twMerge('checkbox', className, clsx({
    [`checkbox-${size}`]: size,
    [`checkbox-${color}`]: color
  }));
  const checkboxRef = useRef(null);
  useImperativeHandle(ref, () => checkboxRef.current);
  useEffect(() => {
    if (!checkboxRef.current) {
      return;
    }

    if (indeterminate) {
      checkboxRef.current.indeterminate = true;
    } else {
      checkboxRef.current.indeterminate = false;
    }
  }, [indeterminate]);
  return jsx("input", _extends$1({}, props, {
    ref: checkboxRef,
    type: "checkbox",
    checked: checked,
    defaultChecked: defaultChecked,
    "data-theme": dataTheme,
    className: classes
  }));
});
Checkbox.displayName = 'Checkbox';

const _excluded$D = ["value", "placeholder", "bordered", "borderOffset", "size", "color", "dataTheme", "className", "type"];
const Input = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    value,
    placeholder,
    bordered = true,
    borderOffset,
    size,
    color,
    dataTheme,
    className,
    type
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$D);

  const classes = twMerge('input', className, clsx({
    [`input-${size}`]: size,
    [`input-${color}`]: color,
    [`focus:outline-offset-0`]: !borderOffset,
    'input-bordered': bordered
  }));
  return jsx("input", _extends$1({}, props, {
    ref: ref,
    type: type,
    value: value,
    placeholder: placeholder,
    "data-theme": dataTheme,
    className: classes
  }));
});
Input.displayName = 'Input';

const _excluded$C = ["color", "size", "name", "dataTheme", "className"];
const Radio = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    color,
    size,
    name,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$C);

  const classes = twMerge('radio', className, clsx({
    [`radio-${size}`]: size,
    [`radio-${color}`]: color
  }));
  return jsx("input", _extends$1({}, props, {
    ref: ref,
    type: "radio",
    name: name,
    "data-theme": dataTheme,
    className: classes
  }));
});
Radio.displayName = 'Radio';

const _excluded$B = ["color", "size", "step", "dataTheme", "className"];
const Range = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    color,
    size,
    step,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$B);

  const classes = twMerge('range', className, clsx({
    [`range-${size}`]: size,
    [`range-${color}`]: color
  }));

  const isNumeric = n => !isNaN(parseFloat(n)) && isFinite(n);

  const numSteps = useMemo(() => {
    var _Math$ceil;

    const safeStep = Math.max(1, Number(step));
    return (_Math$ceil = Math.ceil(100 / safeStep)) != null ? _Math$ceil : 1;
  }, [props.max, step]);
  return jsxs(Fragment, {
    children: [jsx("input", _extends$1({}, props, {
      ref: ref,
      type: "range",
      step: step,
      "data-theme": dataTheme,
      className: classes
    })), isNumeric(step) && jsx("div", {
      className: "w-full flex justify-between text-xs px-2",
      children: [...Array(numSteps + 1)].map(() => {
        return jsx("span", {
          children: "|"
        });
      })
    })]
  });
});
Range.displayName = 'Range';

const RatingItem = _ref => {
  let props = _extends$1({}, _ref);

  return jsx("input", _extends$1({}, props, {
    type: "radio"
  }));
};

const _excluded$A = ["children", "size", "half", "hidden", "dataTheme", "className"];
const Rating = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    size,
    half,
    hidden,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$A);

  const classes = twMerge('rating', className, clsx({
    [`rating-${size}`]: size,
    'rating-half': half,
    'rating-hidden': hidden
  }));
  return jsx("div", _extends$1({
    "aria-label": "Rating"
  }, props, {
    ref: ref,
    "data-theme": dataTheme,
    className: classes,
    children: children
  }));
});
var Rating$1 = Object.assign(Rating, {
  Item: RatingItem
});

const _excluded$z = ["selectedValue", "value", "children"];

const SelectOption = _ref => {
  let {
    selectedValue,
    value,
    children
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$z);

  return jsx("option", _extends$1({}, props, {
    value: value,
    selected: value === selectedValue,
    children: children
  }));
};

const _excluded$y = ["children", "initialValue", "value", "onChange", "size", "color", "bordered", "borderOffset", "dataTheme", "className"];

const SelectInner = (props, ref) => {
  const {
    children,
    initialValue,
    value,
    onChange,
    size,
    color,
    bordered = true,
    borderOffset,
    dataTheme,
    className
  } = props,
        rest = _objectWithoutPropertiesLoose(props, _excluded$y);

  const classes = twMerge('select', className, clsx({
    [`select-${size}`]: size,
    [`select-${color}`]: color,
    [`focus:outline-offset-0`]: !borderOffset,
    'select-bordered': bordered
  }));
  const [selectedValue, setSelectedValue] = useState(value || initialValue);
  return jsx("select", _extends$1({}, rest, {
    ref: ref,
    "data-theme": dataTheme,
    className: classes,
    onChange: e => {
      setSelectedValue(e.currentTarget.value);
      onChange && onChange(e.currentTarget.value);
    },
    children: children.map(child => {
      return /*#__PURE__*/cloneElement(child, {
        selectedValue: selectedValue
      });
    })
  }));
}; // Make forwardRef work with generic component


const Select = /*#__PURE__*/React.forwardRef(SelectInner);
var Select$1 = Object.assign(Select, {
  Option: SelectOption
});

const _excluded$x = ["color", "size", "dataTheme", "className"];
const Toggle = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    color,
    size,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$x);

  const classes = twMerge('toggle', className, clsx({
    [`toggle-${size}`]: size,
    [`toggle-${color}`]: color
  }));
  return jsx("input", _extends$1({}, props, {
    ref: ref,
    type: "checkbox",
    "data-theme": dataTheme,
    className: classes
  }));
});
Toggle.displayName = 'Toggle';

const _excluded$w = ["bordered", "borderOffset", "color", "dataTheme", "className"];
const Textarea = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    bordered = true,
    borderOffset,
    color,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$w);

  const classes = twMerge('textarea', className, clsx({
    [`textarea-${color}`]: color,
    [`focus:outline-offset-0`]: !borderOffset,
    'textarea-bordered': bordered
  }));
  return jsx("textarea", _extends$1({}, props, {
    "data-theme": dataTheme,
    className: classes,
    ref: ref
  }));
});

const _excluded$v = ["children", "size", "horizontal", "dataTheme", "className"];
const Artboard = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    children,
    size,
    horizontal,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$v);

  const classes = twMerge('artboard', 'artboard-demo', className, clsx({
    [`phone-${size}`]: size,
    horizontal: horizontal
  }));
  return jsx("div", _extends$1({
    "aria-label": "Artboard"
  }, props, {
    ref: ref,
    "data-theme": dataTheme,
    className: classes,
    children: children
  }));
});
Artboard.displayName = 'Artboard';

const _excluded$u = ["vertical", "children", "dataTheme", "className"];
const ButtonGroup = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    vertical,
    children,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$u);

  const classes = twMerge('btn-group', className, clsx({
    'btn-group-vertical': vertical
  }));
  return jsx("div", _extends$1({
    "aria-label": `Group of ${children.length} buttons`
  }, props, {
    "data-theme": dataTheme,
    className: classes,
    ref: ref,
    children: children
  }));
});

const _excluded$t = ["children", "vertical", "dataTheme", "className"];

const Divider = _ref => {
  let {
    children,
    vertical,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$t);

  const classes = twMerge('divider', className, clsx({
    'divider-vertical': vertical
  }));
  return jsx("div", _extends$1({
    role: "separator"
  }, props, {
    "data-theme": dataTheme,
    className: classes,
    children: children
  }));
};

const _excluded$s = ["children", "side", "open", "mobile", "end", "dataTheme", "className", "toggleClassName", "contentClassName", "sideClassName", "overlayClassName", "onClickOverlay"];

const Drawer = _ref => {
  let {
    children,
    side,
    open,
    mobile,
    end,
    dataTheme,
    className,
    toggleClassName,
    contentClassName,
    sideClassName,
    overlayClassName,
    onClickOverlay
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$s);

  const classes = twMerge('drawer', className, clsx({
    'drawer-mobile': mobile,
    'drawer-end': end
  }));
  return jsxs("div", _extends$1({
    "aria-expanded": open
  }, props, {
    "data-theme": dataTheme,
    className: classes,
    children: [jsx("input", {
      type: "checkbox",
      className: clsx('drawer-toggle', toggleClassName),
      checked: open,
      readOnly: true
    }), jsx("div", {
      className: clsx('drawer-content', contentClassName),
      children: children
    }), jsxs("div", {
      className: clsx('drawer-side', sideClassName),
      children: [jsx("label", {
        className: clsx('drawer-overlay', overlayClassName),
        onClick: onClickOverlay
      }), side]
    })]
  }));
};

const _excluded$r = ["className"];
const FooterTitle = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$r);

  const classes = twMerge('footer-title', className);
  return jsx("span", _extends$1({}, props, {
    className: classes,
    ref: ref
  }));
});

const _excluded$q = ["center", "dataTheme", "className"];
const Footer = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    center,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$q);

  const classes = twMerge('footer', className, clsx({
    'footer-center': center
  }));
  return jsx("div", _extends$1({
    role: "contentinfo"
  }, props, {
    "data-theme": dataTheme,
    className: classes,
    ref: ref
  }));
});
var Footer$1 = Object.assign(Footer, {
  Title: FooterTitle
});

const _excluded$p = ["dataTheme", "className", "children"];
const HeroContent = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    dataTheme,
    className,
    children
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$p);

  const classes = twMerge('hero-content', className);
  return jsx("div", _extends$1({}, props, {
    "data-theme": dataTheme,
    className: classes,
    ref: ref,
    children: children
  }));
});

const _excluded$o = ["dataTheme", "className", "children"];
const HeroOverlay = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    dataTheme,
    className,
    children
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$o);

  const classes = twMerge('hero-overlay', className);
  return jsx("div", _extends$1({}, props, {
    "data-theme": dataTheme,
    className: classes,
    ref: ref,
    children: children
  }));
});

const _excluded$n = ["dataTheme", "className", "children"];
const Hero = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    dataTheme,
    className,
    children
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$n);

  const classes = twMerge('hero', className);
  return jsx("div", _extends$1({
    role: "banner"
  }, props, {
    "data-theme": dataTheme,
    className: classes,
    ref: ref,
    children: children
  }));
});
var Hero$1 = Object.assign(Hero, {
  Content: HeroContent,
  Overlay: HeroOverlay
});

const _excluded$m = ["children", "item", "horizontal", "vertical", "dataTheme", "className", "innerRef"];
const Indicator = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    item,
    horizontal = 'end',
    vertical = 'top',
    dataTheme,
    className,
    innerRef
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$m);

  const classes = twMerge('indicator-item', className, clsx({
    [`indicator-${horizontal}`]: horizontal,
    [`indicator-${vertical}`]: vertical
  }));
  return jsxs("div", {
    "data-theme": dataTheme,
    className: "indicator",
    ref: ref,
    children: [jsx("div", _extends$1({
      "aria-label": "Indicator"
    }, props, {
      className: classes,
      ref: innerRef,
      children: item
    })), children]
  });
});

const _excluded$l = ["children", "size", "vertical", "dataTheme", "className"];
const InputGroup = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    children,
    size,
    vertical,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$l);

  const classes = twMerge('input-group', className, clsx({
    [`input-group-${size}`]: size,
    'input-group-vertical': vertical
  }));
  return jsx("label", _extends$1({}, props, {
    "data-theme": dataTheme,
    className: classes,
    ref: ref,
    children: children
  }));
});
InputGroup.displayName = 'InputGroup';

const _excluded$k = ["src", "variant", "dataTheme", "className"];
const Mask = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    src,
    variant,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$k);

  const classes = twMerge('mask', className, clsx({
    [`mask-${variant}`]: variant
  }));
  return jsx("img", _extends$1({}, props, {
    "data-theme": dataTheme,
    className: classes,
    src: src,
    ref: ref
  }));
});

const _excluded$j = ["dataTheme", "className", "children"];
const Stack = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    dataTheme,
    className,
    children
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$j);

  const classes = twMerge('stack', className);
  return jsx("div", _extends$1({
    "aria-label": "Stack"
  }, props, {
    ref: ref,
    "data-theme": dataTheme,
    className: classes,
    children: children
  }));
});
Stack.displayName = 'Stack';

const _excluded$i = ["horizontal", "vertical", "className", "children"];
const horizontalOptions = {
  start: 'toast-start',
  center: 'toast-center',
  end: 'toast-end'
};
const verticalOptions = {
  top: 'toast-top',
  middle: 'toast-middle',
  bottom: 'toast-bottom'
};
const Toast = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    horizontal = 'end',
    vertical = 'bottom',
    className,
    children
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$i);

  return jsx("div", _extends$1({}, props, {
    className: twMerge('toast', horizontalOptions[horizontal], verticalOptions[vertical], className),
    ref: ref,
    children: children
  }));
});
Toast.displayName = 'Toast';

const _excluded$h = ["children", "href"];
const BreadcrumbsItem = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    href
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$h);

  return jsx("li", _extends$1({
    role: "link"
  }, props, {
    ref: ref,
    children: href ? jsx("a", {
      href: href,
      children: children
    }) : jsx(Fragment, {
      children: children
    })
  }));
});

const _excluded$g = ["children", "dataTheme", "className", "innerProps", "innerRef"];
const Breadcrumbs = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    dataTheme,
    className,
    innerProps,
    innerRef
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$g);

  const classes = twMerge('breadcrumbs', 'text-sm', className);
  return jsx("div", _extends$1({
    role: "navigation",
    "aria-label": "Breadcrumbs"
  }, props, {
    "data-theme": dataTheme,
    className: classes,
    ref: ref,
    children: jsx("ul", _extends$1({}, innerProps, {
      ref: innerRef,
      children: children
    }))
  }));
});
var Breadcrumbs$1 = Object.assign(Breadcrumbs, {
  Item: BreadcrumbsItem
});

const _excluded$f = ["children", "className"];
const BottomNavigationLabel = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    children,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$f);

  const classes = twMerge('btm-nav-label', className);
  return jsx("span", _extends$1({}, props, {
    className: classes,
    ref: ref,
    children: children
  }));
});

const _excluded$e = ["size", "dataTheme", "className", "children"];
const BottomNavigation = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    size,
    dataTheme,
    className,
    children
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$e);

  const classes = twMerge('btm-nav', clsx({
    [`btm-nav-${size}`]: size
  }), className);
  return jsx("div", _extends$1({}, props, {
    "data-theme": dataTheme,
    className: classes,
    ref: ref,
    children: children
  }));
});
BottomNavigation.displayName = 'BottomNavigation';
var BottomNavigation$1 = Object.assign(BottomNavigation, {
  Label: BottomNavigationLabel
});

const _excluded$d = ["children", "href", "color", "hover", "dataTheme", "className"];
const Link = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    children,
    href,
    color,
    hover = true,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$d);

  const classes = twMerge('link', className, clsx({
    [`link-${color}`]: color,
    'link-hover': hover
  }));
  return jsx("a", _extends$1({}, props, {
    href: href,
    "data-theme": dataTheme,
    className: classes,
    ref: ref,
    children: children
  }));
});
Link.displayName = 'Link';

const _excluded$c = ["className"];
const MenuTitle = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$c);

  const classes = twMerge('menu-title', className);
  return jsx("li", _extends$1({}, props, {
    className: classes,
    ref: ref
  }));
});

const _excluded$b = ["className", "disabled"];
const MenuItem = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    className,
    disabled
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$b);

  const classes = twMerge(className, clsx({
    disabled: disabled
  }));
  return jsx("li", _extends$1({
    role: "menuitem",
    className: classes
  }, props, {
    ref: ref
  }));
});

const _excluded$a = ["normal", "compact", "horizontal", "vertical", "dataTheme", "className"];
const DYNAMIC_MODIFIERS = {
  compact: {
    true: 'menu-compact',
    xs: 'xs:menu-compact',
    sm: 'sm:menu-compact',
    md: 'md:menu-compact',
    lg: 'lg:menu-compact'
  },
  normal: {
    true: 'menu-normal',
    xs: 'xs:menu-normal',
    sm: 'sm:menu-normal',
    md: 'md:menu-normal',
    lg: 'lg:menu-normal'
  },
  horizontal: {
    true: 'menu-horizontal',
    xs: 'xs:menu-horizontal',
    sm: 'sm:menu-horizontal',
    md: 'md:menu-horizontal',
    lg: 'lg:menu-horizontal'
  },
  vertical: {
    true: 'menu-vertical',
    xs: 'xs:menu-vertical',
    sm: 'sm:menu-vertical',
    md: 'md:menu-vertical',
    lg: 'lg:menu-vertical'
  }
};
const Menu = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    normal,
    compact,
    horizontal,
    vertical,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$a);

  const classes = twMerge('menu', className, clsx({
    [compact && DYNAMIC_MODIFIERS.compact[compact.toString()] || '']: compact,
    [normal && DYNAMIC_MODIFIERS.normal[normal.toString()] || '']: normal,
    [horizontal && DYNAMIC_MODIFIERS.horizontal[horizontal.toString()] || '']: horizontal,
    [vertical && DYNAMIC_MODIFIERS.vertical[vertical.toString()] || '']: vertical
  }));
  return jsx("ul", _extends$1({
    role: "menu",
    "data-theme": dataTheme,
    className: classes
  }, props, {
    ref: ref
  }));
});
var Menu$1 = Object.assign(Menu, {
  Title: MenuTitle,
  Item: MenuItem
});

const NavbarSection = /*#__PURE__*/React.forwardRef(({
  children,
  section,
  dataTheme,
  className,
  style
}, ref) => {
  const classes = twMerge(className, clsx({
    ['flex-1']: section !== 'center',
    [`navbar-${section}`]: section
  }));
  return jsx("div", {
    "data-theme": dataTheme,
    className: classes,
    style: style,
    ref: ref,
    children: children
  });
});

const _excluded$9 = ["children", "dataTheme", "className"];
const Navbar = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$9);

  const classes = twMerge('navbar', className);
  return jsx("div", _extends$1({
    role: "navigation",
    "aria-label": "Navbar"
  }, props, {
    "data-theme": dataTheme,
    className: classes,
    ref: ref,
    children: children
  }));
});
const NavbarStart = /*#__PURE__*/React.forwardRef((props, ref) => jsx(NavbarSection, _extends$1({}, props, {
  section: "start",
  ref: ref
})));
const NavbarCenter = /*#__PURE__*/React.forwardRef((props, ref) => jsx(NavbarSection, _extends$1({}, props, {
  section: "center",
  ref: ref
})));
const NavbarEnd = /*#__PURE__*/React.forwardRef((props, ref) => jsx(NavbarSection, _extends$1({}, props, {
  section: "end",
  ref: ref
})));
var Navbar$1 = Object.assign(Navbar, {
  Start: NavbarStart,
  Center: NavbarCenter,
  End: NavbarEnd
});

const Pagination = ButtonGroup;

const _excluded$8 = ["children", "value", "color", "dataTheme", "className"];
const Step = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    value,
    color,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$8);

  const classes = twMerge('step', className, clsx({
    [`step-${color}`]: color
  }));
  return jsx("li", _extends$1({
    "aria-label": "Step"
  }, props, {
    "data-theme": dataTheme,
    "data-content": value,
    className: classes,
    ref: ref,
    children: children
  }));
});

const _excluded$7 = ["children", "dataTheme", "className", "vertical", "horizontal"];
const Steps = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    dataTheme,
    className,
    vertical,
    horizontal
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$7);

  const classes = twMerge('steps', className, clsx({
    ['steps-vertical']: vertical,
    ['steps-horizontal']: horizontal
  }));
  return jsx("ul", _extends$1({
    "aria-label": "Steps",
    role: "group"
  }, props, {
    "data-theme": dataTheme,
    className: classes,
    ref: ref,
    children: children
  }));
});
var Steps$1 = Object.assign(Steps, {
  Step
});

const _excluded$6 = ["children", "value", "activeValue", "onClick", "size", "variant", "className", "style"];

const TabInner = (_ref, ref) => {
  let {
    children,
    value,
    activeValue,
    onClick,
    size,
    variant,
    className,
    style
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$6);

  const classes = twMerge('tab', className, clsx({
    'tab-active': value != null && value === activeValue,
    [`tab-${size}`]: size,
    [`tab-${variant}`]: variant
  }));
  return jsx("a", _extends$1({
    role: "tab"
  }, props, {
    className: classes,
    style: style,
    onClick: () => onClick && onClick(value),
    children: children
  }));
}; // Make forwardRef work with generic component


const Tab = /*#__PURE__*/React.forwardRef(TabInner);

const _excluded$5 = ["children", "value", "onChange", "variant", "size", "boxed", "dataTheme", "className"];

const TabsInner = (_ref, ref) => {
  let {
    children,
    value,
    onChange,
    variant,
    size,
    boxed,
    dataTheme,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$5);

  const [activeValue, setActiveValue] = useState(value);
  const classes = twMerge('tabs', className, clsx({
    'tabs-boxed': boxed
  }));
  return jsx("div", _extends$1({
    ref: ref,
    role: "tablist"
  }, props, {
    "data-theme": dataTheme,
    className: classes,
    children: children.map(child => {
      return /*#__PURE__*/cloneElement(child, {
        variant,
        size,
        activeValue: activeValue,
        onClick: value => {
          setActiveValue(value);
          onChange && onChange(value);
        }
      });
    })
  }));
}; // Make forwardRef work with generic component


const Tabs = /*#__PURE__*/React.forwardRef(TabsInner);
var Tabs$1 = Object.assign(Tabs, {
  Tab
});

const _excluded$4 = ["dataPrefix", "dataTheme", "status", "className", "children", "innerProps", "innerRef"];
const CodeMockupLine = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    dataPrefix,
    dataTheme,
    status,
    className,
    children,
    innerProps,
    innerRef
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$4);

  const classes = twMerge(clsx({
    [`bg-${status}`]: status,
    [`text-${status}-content`]: status
  }), className);

  const allProps = _extends$1({}, props, {
    className: classes
  }, dataPrefix !== false && {
    'data-prefix': dataPrefix || '>'
  });

  return jsx("pre", _extends$1({}, allProps, {
    "data-theme": dataTheme,
    className: classes,
    ref: ref,
    children: jsx("code", _extends$1({}, innerProps, {
      ref: innerRef,
      children: children
    }))
  }));
});
CodeMockupLine.displayName = 'CodeMockup.Line';

const _excluded$3 = ["dataTheme", "className", "children"];
const CodeMockup = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    dataTheme,
    className,
    children
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$3);

  const classes = twMerge('mockup-code', className);
  return jsx("div", _extends$1({
    "aria-label": "Code mockup"
  }, props, {
    "data-theme": dataTheme,
    className: classes,
    ref: ref,
    children: children
  }));
});
CodeMockup.displayName = 'CodeMockup';
const CodeMockupNamespace = Object.assign(CodeMockup, {
  Line: CodeMockupLine
});

const _excluded$2 = ["color", "dataTheme", "className", "children", "innerRef", "innerProps"];
const PhoneMockup = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    color,
    dataTheme,
    className,
    children,
    innerRef,
    innerProps
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$2);

  const classes = twMerge('mockup-phone', clsx({
    [`border-${color}`]: color
  }), className);
  return jsxs("div", _extends$1({
    "aria-label": "Phone mockup"
  }, props, {
    "data-theme": dataTheme,
    className: classes,
    ref: ref,
    children: [jsx("div", {
      className: "camera"
    }), jsx("div", {
      className: "display",
      children: jsx("div", _extends$1({}, innerProps, {
        className: twMerge('artboard artboard-demo phone-1', innerProps == null ? void 0 : innerProps.className),
        ref: innerRef,
        children: children
      }))
    })]
  }));
});
PhoneMockup.displayName = 'PhoneMockup';

const defaultTheme = 'light';

const _excluded$1 = ["border", "borderColor", "backgroundColor", "frameColor", "dataTheme", "className", "children"];
const WindowMockup = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    border,
    borderColor,
    backgroundColor,
    frameColor,
    dataTheme,
    className,
    children
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded$1);

  // Set border color to framecolor or 'bg-base-300', if border color is not defined
  const borderColorValue = borderColor ? borderColor : frameColor || 'base-300';
  const classes = twMerge('mockup-window', border && `border border-${borderColorValue}`, clsx({
    [`border-${borderColorValue}`]: borderColorValue,
    [`bg-${frameColor}`]: frameColor
  }), className);
  console.log('classes', classes); // If border is true, then we need to add the border-t and padding classes to the children
  // if more than one child is passed in, or the single child is not a valid element, then we need to wrap the child/children in a div

  const numChildren = React.Children.count(children);
  const firstChild = numChildren > 0 && React.Children.toArray(children)[0]; // List of classes that child element will have

  const innerClasses = clsx(backgroundColor && `bg-${backgroundColor}`, border && `border-t border-${borderColorValue}`, 'p-4');
  console.log('innerClasses', innerClasses); // Add the innerClasses to the child element, merging classNames if existing, or wrapping with div and adding innerClasses

  const innerEl = firstChild && /*#__PURE__*/React.isValidElement(firstChild) ? /*#__PURE__*/React.cloneElement(firstChild, {
    className: twMerge(innerClasses, firstChild.props.className)
  }) : jsx("div", {
    className: innerClasses,
    children: children
  });
  return jsx("div", _extends$1({
    "aria-label": "Window mockup"
  }, props, {
    "data-theme": dataTheme,
    className: classes,
    ref: ref,
    children: innerEl
  }));
});
WindowMockup.displayName = 'WindowMockup';

const ThemeContext = /*#__PURE__*/React.createContext({
  theme: 'light',
  setTheme: () => {}
});

const getThemeFromClosestAncestor = ref => {
  if (!ref.current) return;
  const matches = ref.current.closest('[data-theme]');
  if (matches) return matches.getAttribute('data-theme');
};

const _excluded = ["children", "dataTheme", "onChange", "className"];
const Theme = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    dataTheme,
    onChange,
    className
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded);

  // Either use provided ref or create a new ref
  const themeRef = useRef(ref == null ? void 0 : ref.current);
  const closestAncestorTheme = getThemeFromClosestAncestor(themeRef); // If no theme is provided, use the closest ancestor theme, if no ancestor theme, fallback to default theme (defined in constants)

  const [theme, setTheme] = useState(dataTheme || closestAncestorTheme || defaultTheme);

  const handleThemeChange = theme => {
    // Fire custom onChange, if provided. ie, user provided function to store theme in session/local storage
    onChange && onChange(theme); // Update state/context

    setTheme(theme);
  }; // Properly handle changes to theme prop on Theme component


  useEffect(() => {
    if (dataTheme !== theme) {
      dataTheme && handleThemeChange(dataTheme);
    }
  }, [dataTheme]);
  return jsx(ThemeContext.Provider, {
    value: {
      theme,
      setTheme: handleThemeChange
    },
    children: jsx("div", _extends$1({}, props, {
      "data-theme": theme,
      className: className,
      ref: themeRef,
      children: children
    }))
  });
});

const useTheme = value => {
  const {
    theme,
    setTheme
  } = useContext(ThemeContext);
  useEffect(() => {
    if (value && theme !== value) {
      setTheme(value);
    }
  }, [value]);
  return {
    theme,
    setTheme
  };
};

export { Alert, Artboard, Avatar$1 as Avatar, Badge, BottomNavigation$1 as BottomNavigation, Breadcrumbs$1 as Breadcrumbs, Button, ButtonGroup, Card$1 as Card, Carousel$1 as Carousel, Checkbox, CodeMockupNamespace as CodeMockup, Collapse$1 as Collapse, Countdown, Divider, Drawer, Dropdown$1 as Dropdown, Footer$1 as Footer, Form$1 as Form, Hero$1 as Hero, Indicator, Input, InputGroup, Kbd, Link, Mask, Menu$1 as Menu, Modal$1 as Modal, Navbar$1 as Navbar, Pagination, PhoneMockup, Progress, RadialProgress, Radio, Range, Rating$1 as Rating, Select$1 as Select, Stack, Stats$1 as Stats, Steps$1 as Steps, Swap, Table$1 as Table, Tabs$1 as Tabs, Textarea, Theme, Toast, Toggle, Tooltip, WindowMockup, useTheme };
